// @braintrust/core/typespecs-stainless/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// typespecs/index.ts
var index_exports = {};
__export(index_exports, {
  BRAINTRUST_ATTACHMENT: () => BRAINTRUST_ATTACHMENT,
  BRAINTRUST_PARAMS: () => BRAINTRUST_PARAMS,
  DEFAULT_IF_EXISTS: () => DEFAULT_IF_EXISTS,
  EXTERNAL_ATTACHMENT: () => EXTERNAL_ATTACHMENT,
  aclBatchUpdateRequestSchema: () => aclBatchUpdateRequestSchema,
  aclBatchUpdateResponseSchema: () => aclBatchUpdateResponseSchema,
  aclItemSchema: () => aclItemSchema,
  aclObjectTypeEnum: () => aclObjectTypeEnum,
  aclSchema: () => aclSchema,
  aiSecretSchema: () => aiSecretSchema,
  apiKeySchema: () => apiKeySchema,
  apiSpecEventObjectSchemas: () => apiSpecEventObjectSchemas,
  apiSpecObjectSchemas: () => apiSpecObjectSchemas,
  appLimitParamSchema: () => appLimitParamSchema,
  asyncScoringControlSchema: () => asyncScoringControlSchema,
  asyncScoringStateSchema: () => asyncScoringStateSchema,
  attachmentReferenceSchema: () => attachmentReferenceSchema,
  attachmentStatusSchema: () => attachmentStatusSchema,
  baseSSEEventSchema: () => baseSSEEventSchema,
  braintrustAttachmentReferenceSchema: () => braintrustAttachmentReferenceSchema,
  callEventSchema: () => callEventSchema,
  chatCompletionContentPartImageSchema: () => chatCompletionContentPartImageSchema,
  chatCompletionContentPartSchema: () => chatCompletionContentPartSchema,
  chatCompletionContentPartTextSchema: () => chatCompletionContentPartTextSchema,
  chatCompletionMessageParamSchema: () => chatCompletionMessageParamSchema,
  chatCompletionMessageToolCallSchema: () => chatCompletionMessageToolCallSchema,
  codeBundleSchema: () => codeBundleSchema,
  comparisonExperimentIdParamSchema: () => comparisonExperimentIdParamSchema,
  createAISecretSchema: () => createAISecretSchema,
  createApiKeyOutputSchema: () => createApiKeyOutputSchema,
  createApiKeySchema: () => createApiKeySchema,
  createDatasetSchema: () => createDatasetSchema,
  createEnvVarSchema: () => createEnvVarSchema,
  createExperimentSchema: () => createExperimentSchema,
  createFunctionSchema: () => createFunctionSchema,
  createGroupSchema: () => createGroupSchema,
  createProjectSchema: () => createProjectSchema,
  createProjectScoreSchema: () => createProjectScoreSchema,
  createProjectTagSchema: () => createProjectTagSchema,
  createPromptSchema: () => createPromptSchema,
  createSpanIframeSchema: () => createSpanIframeSchema,
  createViewSchema: () => createViewSchema,
  crossObjectInsertRequestSchema: () => crossObjectInsertRequestSchema,
  crossObjectInsertResponseSchema: () => crossObjectInsertResponseSchema,
  customColumnSchema: () => customColumnSchema,
  customTypes: () => customTypes,
  customTypesMode: () => customTypesMode,
  datasetEventSchema: () => datasetEventSchema,
  datasetSchema: () => datasetSchema,
  datetimeStringSchema: () => datetimeStringSchema,
  deleteAISecretSchema: () => deleteAISecretSchema,
  endingBeforeSchema: () => endingBeforeSchema,
  envVarObjectTypeEnum: () => envVarObjectTypeEnum,
  envVarSchema: () => envVarSchema,
  eventObjectType: () => eventObjectType,
  experimentEventSchema: () => experimentEventSchema,
  experimentSchema: () => experimentSchema,
  extendedSavedFunctionIdSchema: () => extendedSavedFunctionIdSchema,
  externalAttachmentReferenceSchema: () => externalAttachmentReferenceSchema,
  feedbackResponseSchema: () => feedbackResponseSchema,
  fetchEventsRequestSchema: () => fetchEventsRequestSchema,
  fetchLimitParamSchema: () => fetchLimitParamSchema,
  fetchPaginationCursorSchema: () => fetchPaginationCursorSchema,
  functionDataSchema: () => functionDataSchema,
  functionFormatEnum: () => functionFormatEnum,
  functionIdSchema: () => functionIdSchema,
  functionObjectTypeEnum: () => functionObjectTypeEnum,
  functionOutputTypeEnum: () => functionOutputTypeEnum,
  functionSchema: () => functionSchema2,
  functionTypeEnum: () => functionTypeEnum,
  getEventObjectDescription: () => getEventObjectDescription,
  getEventObjectType: () => getEventObjectType,
  getObjectArticle: () => getObjectArticle,
  gitFieldsSchema: () => gitFieldsSchema,
  gitMetadataSettingsSchema: () => gitMetadataSettingsSchema,
  groupSchema: () => groupSchema,
  ifExistsEnum: () => ifExistsEnum,
  insertEventsResponseSchema: () => insertEventsResponseSchema,
  invokeApiSchema: () => invokeApiSchema,
  invokeFunctionNonIdArgsSchema: () => invokeFunctionNonIdArgsSchema,
  invokeFunctionSchema: () => invokeFunctionSchema,
  invokeParent: () => invokeParent,
  jsonSchema: () => jsonSchema,
  literalSchema: () => literalSchema,
  makeObjectIdsFilterSchema: () => makeObjectIdsFilterSchema,
  maxOverWindowSchema: () => maxOverWindowSchema,
  maxRootSpanIdSchema: () => maxRootSpanIdSchema,
  maxXactIdSchema: () => maxXactIdSchema,
  memberSchema: () => memberSchema,
  messageRoleSchema: () => messageRoleSchema,
  modelParamsSchema: () => modelParamsSchema,
  objectReferenceSchema: () => objectReferenceSchema,
  objectTypeSummarizeResponseSchemas: () => objectTypeSummarizeResponseSchemas,
  objectTypes: () => objectTypes,
  objectTypesWithEvent: () => objectTypesWithEvent,
  onlineScoreConfigSchema: () => onlineScoreConfigSchema,
  organizationMembersSchema: () => organizationMembersSchema,
  organizationSchema: () => organizationSchema,
  patchAISecretSchema: () => patchAISecretSchema,
  patchDatasetSchema: () => patchDatasetSchema,
  patchEnvVarSchema: () => patchEnvVarSchema,
  patchExperimentSchema: () => patchExperimentSchema,
  patchGroupSchema: () => patchGroupSchema,
  patchOrganizationMembersOutputSchema: () => patchOrganizationMembersOutputSchema,
  patchOrganizationMembersSchema: () => patchOrganizationMembersSchema,
  patchOrganizationSchema: () => patchOrganizationSchema,
  patchProjectSchema: () => patchProjectSchema,
  patchProjectScoreSchema: () => patchProjectScoreSchema,
  patchProjectTagSchema: () => patchProjectTagSchema,
  patchPromptSchema: () => patchPromptSchema,
  patchRoleSchema: () => patchRoleSchema,
  patchSpanIframeSchema: () => patchSpanIframeSchema,
  patchViewSchema: () => patchViewSchema,
  permissionEnum: () => permissionEnum,
  playgroundLogsLogIdLiteralSchema: () => playgroundLogsLogIdLiteralSchema,
  projectLogsEventSchema: () => projectLogsEventSchema,
  projectLogsLogIdLiteralSchema: () => projectLogsLogIdLiteralSchema,
  projectSchema: () => projectSchema,
  projectScoreCategory: () => projectScoreCategory,
  projectScoreSchema: () => projectScoreSchema,
  projectScoreTypeEnum: () => projectScoreTypeEnum,
  projectSettingsSchema: () => projectSettingsSchema,
  projectTagSchema: () => projectTagSchema,
  promptBaseSchema: () => promptBaseSchema,
  promptBlockDataSchema: () => promptBlockDataSchema,
  promptDataSchema: () => promptDataSchema,
  promptLogIdLiteralSchema: () => promptLogIdLiteralSchema,
  promptOptionsSchema: () => promptOptionsSchema,
  promptSchema: () => promptSchema,
  promptSessionEventSchema: () => promptSessionEventSchema,
  promptSessionSchema: () => promptSessionSchema,
  repoInfoSchema: () => repoInfoSchema,
  resourcesSchema: () => resourcesSchema,
  responseFormatJsonSchemaSchema: () => responseFormatJsonSchemaSchema,
  responseFormatSchema: () => responseFormatSchema,
  roleSchema: () => roleSchema,
  runEvalSchema: () => runEvalSchema,
  runtimeContextSchema: () => runtimeContextSchema,
  savedFunctionIdSchema: () => savedFunctionIdSchema,
  scoreSchema: () => scoreSchema,
  spanAttributesSchema: () => spanAttributesSchema,
  spanFieldOrderItem: () => spanFieldOrderItem,
  spanIframeBaseSchema: () => spanIframeBaseSchema,
  spanIframeSchema: () => spanIframeSchema,
  spanTypeSchema: () => spanTypeSchema,
  sseConsoleEventDataSchema: () => sseConsoleEventDataSchema,
  sseConsoleEventSchema: () => sseConsoleEventSchema,
  sseDataEventSchema: () => sseDataEventSchema,
  sseDoneEventSchema: () => sseDoneEventSchema,
  sseErrorEventSchema: () => sseErrorEventSchema,
  sseProgressEventDataSchema: () => sseProgressEventDataSchema,
  sseProgressEventSchema: () => sseProgressEventSchema,
  sseStartEventSchema: () => sseStartEventSchema,
  sseTextEventSchema: () => sseTextEventSchema,
  startingAfterSchema: () => startingAfterSchema,
  streamingModeEnum: () => streamingModeEnum,
  strictPromptDataSchema: () => strictPromptDataSchema,
  summarizeDataParamSchema: () => summarizeDataParamSchema,
  summarizeScoresParamSchema: () => summarizeScoresParamSchema,
  toolFunctionDefinitionSchema: () => toolFunctionDefinitionSchema,
  toolsSchema: () => toolsSchema,
  uploadStatusSchema: () => uploadStatusSchema,
  useFunctionSchema: () => useFunctionSchema,
  userSchema: () => userSchema,
  validRuntimesEnum: () => validRuntimesEnum,
  versionSchema: () => versionSchema,
  viewDataSchema: () => viewDataSchema,
  viewDataSearchSchema: () => viewDataSearchSchema,
  viewOptionsSchema: () => viewOptionsSchema,
  viewSchema: () => viewSchema,
  viewTypeEnum: () => viewTypeEnum
});
module.exports = __toCommonJS(index_exports);

// typespecs/api_types.ts
var import_zod_to_openapi9 = require("@asteasolutions/zod-to-openapi");
var import_zod12 = require("zod");

// typespecs/app_types.ts
var import_zod_to_openapi8 = require("@asteasolutions/zod-to-openapi");
var import_zod11 = require("zod");

// src/zod_util.ts
var import_zod = require("zod");
function objectNullish(object) {
  return new import_zod.z.ZodObject({
    ...object._def,
    shape: () => Object.fromEntries(
      Object.entries(object.shape).map(([k, v]) => [k, v.nullish()])
    )
  });
}

// typespecs/common_types.ts
var import_zod2 = require("zod");
var import_zod_to_openapi = require("@asteasolutions/zod-to-openapi");
(0, import_zod_to_openapi.extendZodWithOpenApi)(import_zod2.z);
var literalSchema = import_zod2.z.union([
  import_zod2.z.string().openapi({ title: "string" }),
  import_zod2.z.number().openapi({ title: "number" }),
  import_zod2.z.boolean().openapi({ title: "boolean" }),
  import_zod2.z.null().openapi({ title: "null" })
]);
var jsonSchema = import_zod2.z.lazy(
  () => import_zod2.z.union([
    literalSchema,
    import_zod2.z.array(jsonSchema).openapi({ title: "array" }),
    import_zod2.z.record(jsonSchema).openapi({ title: "object" })
  ])
);
var datetimeStringSchema = import_zod2.z.string().transform((x, ctx) => {
  const d = new Date(x);
  if (isNaN(d.getTime())) {
    ctx.addIssue({
      code: import_zod2.z.ZodIssueCode.invalid_string,
      validation: "datetime",
      message: "Invalid datetime"
    });
    return import_zod2.z.NEVER;
  }
  return d.toISOString();
}).openapi({ format: "date-time" });
var objectTypes = import_zod2.z.enum([
  "project",
  "experiment",
  "dataset",
  "prompt",
  "prompt_session",
  "role",
  "group",
  "acl",
  "user",
  "project_score",
  "project_tag",
  "span_iframe",
  "function",
  "view",
  "organization",
  "api_key",
  "ai_secret",
  "env_var"
]);
var objectTypesWithEvent = import_zod2.z.enum([
  "project",
  "experiment",
  "dataset",
  "prompt",
  "function",
  "prompt_session"
]);
var eventObjectType = objectTypesWithEvent.exclude(["project"]).or(import_zod2.z.enum(["project_logs"]));
function getEventObjectType(objectType) {
  return objectType === "project" ? "project_logs" : objectType;
}
function getEventObjectDescription(objectType) {
  return getEventObjectType(objectType).replace("_", " ");
}
function getObjectArticle(objectType) {
  return [
    "acl",
    "api_key",
    "experiment",
    "organization",
    "ai_secret",
    "env_var"
  ].includes(objectType) ? "an" : "a";
}
var objectReferenceSchema = import_zod2.z.object({
  object_type: eventObjectType.describe(
    "Type of the object the event is originating from."
  ),
  object_id: import_zod2.z.string().uuid().describe("ID of the object the event is originating from."),
  id: import_zod2.z.string().describe("ID of the original event."),
  _xact_id: import_zod2.z.string().describe("Transaction ID of the original event."),
  created: import_zod2.z.string().nullish().describe(
    "Created timestamp of the original event. Used to help sort in the UI"
  )
}).describe("Reference to the original object and event this was copied from.").openapi("ObjectReference");

// typespecs/custom_types.ts
var import_zod_to_openapi2 = require("@asteasolutions/zod-to-openapi");
var import_zod3 = require("zod");
(0, import_zod_to_openapi2.extendZodWithOpenApi)(import_zod3.z);
var modeSchema = import_zod3.z.enum(["default", "stainless"]);
var mode = modeSchema.parse(
  typeof process === "undefined" ? "default" : "stainless"
);
var modeToTypes = {
  default: {
    unknown: import_zod3.z.unknown(),
    literalTrue: import_zod3.z.literal(true),
    literalFalse: import_zod3.z.literal(false)
  },
  stainless: {
    // Stainless requires schemas which are completely permissive to be
    // tagged.
    unknown: import_zod3.z.unknown().openapi({ ["x-stainless-any"]: true }),
    // Stainless does not support boolean literals in all SDKs.
    literalTrue: import_zod3.z.boolean(),
    literalFalse: import_zod3.z.boolean()
  }
};
var customTypes = modeToTypes[mode];
var customTypesMode = mode;

// typespecs/prompt.ts
var import_zod_to_openapi5 = require("@asteasolutions/zod-to-openapi");
var import_zod7 = require("zod");

// typespecs/openai/messages.ts
var import_zod_to_openapi3 = require("@asteasolutions/zod-to-openapi");
var import_zod4 = require("zod");
(0, import_zod_to_openapi3.extendZodWithOpenApi)(import_zod4.z);
var messageRoleSchema = import_zod4.z.enum(["system", "user", "assistant", "function", "tool", "model"]).openapi("MessageRole");
var chatCompletionSystemMessageParamSchema = import_zod4.z.object({
  content: import_zod4.z.string().default(""),
  role: import_zod4.z.literal("system"),
  name: import_zod4.z.string().optional()
});
var chatCompletionContentPartTextSchema = import_zod4.z.object({
  text: import_zod4.z.string().default(""),
  type: import_zod4.z.literal("text")
}).openapi("ChatCompletionContentPartText");
var imageURLSchema = import_zod4.z.object({
  url: import_zod4.z.string(),
  detail: import_zod4.z.union([
    import_zod4.z.literal("auto").openapi({ title: "auto" }),
    import_zod4.z.literal("low").openapi({ title: "low" }),
    import_zod4.z.literal("high").openapi({ title: "high" })
  ]).optional()
});
var chatCompletionContentPartImageSchema = import_zod4.z.object({
  image_url: imageURLSchema,
  type: import_zod4.z.literal("image_url")
}).openapi("ChatCompletionContentPartImage");
var chatCompletionContentPartSchema = import_zod4.z.union([
  chatCompletionContentPartTextSchema.openapi({ title: "text" }),
  chatCompletionContentPartImageSchema.openapi({ title: "image_url" })
]).openapi("ChatCompletionContentPart");
var chatCompletionContentSchema = import_zod4.z.union([
  import_zod4.z.string().default("").openapi({ title: "text" }),
  import_zod4.z.array(
    chatCompletionContentPartSchema.openapi({
      title: "chat_completion_content_part"
    })
  ).openapi({ title: "array" })
]).openapi("ChatCompletionContent");
var chatCompletionUserMessageParamSchema = import_zod4.z.object({
  content: chatCompletionContentSchema,
  role: import_zod4.z.literal("user"),
  name: import_zod4.z.string().optional()
});
var functionCallSchema = import_zod4.z.object({
  arguments: import_zod4.z.string(),
  name: import_zod4.z.string()
});
var functionSchema = import_zod4.z.object({
  arguments: import_zod4.z.string(),
  name: import_zod4.z.string()
});
var chatCompletionToolMessageParamSchema = import_zod4.z.object({
  content: import_zod4.z.string().default(""),
  role: import_zod4.z.literal("tool"),
  tool_call_id: import_zod4.z.string().default("")
});
var chatCompletionFunctionMessageParamSchema = import_zod4.z.object({
  content: import_zod4.z.string().default(""),
  name: import_zod4.z.string(),
  role: import_zod4.z.literal("function")
});
var chatCompletionMessageToolCallSchema = import_zod4.z.object({
  id: import_zod4.z.string(),
  function: functionSchema,
  type: import_zod4.z.literal("function")
}).openapi("ChatCompletionMessageToolCall");
var chatCompletionAssistantMessageParamSchema = import_zod4.z.object({
  role: import_zod4.z.literal("assistant"),
  content: import_zod4.z.string().nullish(),
  // NOTE: It's important to keep these optional, rather than nullish, to stay
  // inline with the OpenAI SDK's type definition.
  function_call: functionCallSchema.nullish().transform((x) => x != null ? x : void 0),
  name: import_zod4.z.string().nullish().transform((x) => x != null ? x : void 0),
  tool_calls: import_zod4.z.array(chatCompletionMessageToolCallSchema).nullish().transform((x) => x != null ? x : void 0)
});
var chatCompletionFallbackMessageParamSchema = import_zod4.z.object({
  role: messageRoleSchema.exclude([
    "system",
    "user",
    "assistant",
    "tool",
    "function"
  ]),
  content: import_zod4.z.string().nullish()
});
var chatCompletionOpenAIMessageParamSchema = import_zod4.z.union([
  chatCompletionSystemMessageParamSchema.openapi({ title: "system" }),
  chatCompletionUserMessageParamSchema.openapi({ title: "user" }),
  chatCompletionAssistantMessageParamSchema.openapi({ title: "assistant" }),
  chatCompletionToolMessageParamSchema.openapi({ title: "tool" }),
  chatCompletionFunctionMessageParamSchema.openapi({ title: "function" })
]);
var chatCompletionMessageParamSchema = import_zod4.z.union([
  chatCompletionOpenAIMessageParamSchema.openapi({ title: "openai" }),
  chatCompletionFallbackMessageParamSchema.openapi({ title: "fallback" })
]).openapi("ChatCompletionMessageParam");

// typespecs/function_id.ts
var import_zod_to_openapi4 = require("@asteasolutions/zod-to-openapi");
var import_zod5 = require("zod");
(0, import_zod_to_openapi4.extendZodWithOpenApi)(import_zod5.z);
var savedFunctionIdSchema = import_zod5.z.union([
  import_zod5.z.object({
    type: import_zod5.z.literal("function"),
    id: import_zod5.z.string()
  }).openapi({ title: "function" }),
  import_zod5.z.object({
    type: import_zod5.z.literal("global"),
    name: import_zod5.z.string()
  }).openapi({ title: "global" })
]).openapi("SavedFunctionId");
var extendedSavedFunctionIdSchema = savedFunctionIdSchema.or(
  import_zod5.z.object({
    type: import_zod5.z.literal("slug"),
    project_id: import_zod5.z.string(),
    slug: import_zod5.z.string()
  })
);

// typespecs/openai/tools.ts
var import_zod6 = require("zod");
var functionParametersSchema = import_zod6.z.record(customTypes.unknown);
var functionDefinitionSchema = import_zod6.z.object({
  name: import_zod6.z.string(),
  description: import_zod6.z.string().optional(),
  parameters: functionParametersSchema.optional()
});
var chatCompletionToolSchema = import_zod6.z.object({
  function: functionDefinitionSchema,
  type: import_zod6.z.literal("function")
});
var toolsSchema = import_zod6.z.array(chatCompletionToolSchema);

// typespecs/prompt.ts
(0, import_zod_to_openapi5.extendZodWithOpenApi)(import_zod7.z);
var promptBlockDataSchema = import_zod7.z.union([
  import_zod7.z.object({
    type: import_zod7.z.literal("completion"),
    content: import_zod7.z.string()
  }).openapi({ title: "completion" }),
  import_zod7.z.object({
    type: import_zod7.z.literal("chat"),
    messages: import_zod7.z.array(chatCompletionMessageParamSchema),
    tools: import_zod7.z.string().optional()
  }).openapi({ title: "chat" })
]);
var braintrustModelParamsSchema = import_zod7.z.object({
  use_cache: import_zod7.z.boolean().optional()
});
var BRAINTRUST_PARAMS = Object.keys(braintrustModelParamsSchema.shape);
var responseFormatJsonSchemaSchema = import_zod7.z.object({
  name: import_zod7.z.string(),
  description: import_zod7.z.string().optional(),
  schema: import_zod7.z.union([
    import_zod7.z.record(customTypes.unknown).openapi({ title: "object" }),
    import_zod7.z.string().openapi({ title: "string" })
  ]).optional(),
  strict: import_zod7.z.boolean().nullish()
});
var responseFormatSchema = import_zod7.z.union([
  import_zod7.z.object({ type: import_zod7.z.literal("json_object") }).openapi({ title: "json_object" }),
  import_zod7.z.object({
    type: import_zod7.z.literal("json_schema"),
    json_schema: responseFormatJsonSchemaSchema
  }).openapi({ title: "json_schema" }),
  import_zod7.z.object({ type: import_zod7.z.literal("text") }).openapi({ title: "text" })
]);
var openAIModelParamsSchema = import_zod7.z.object({
  temperature: import_zod7.z.number().optional(),
  top_p: import_zod7.z.number().optional(),
  max_tokens: import_zod7.z.number().optional(),
  max_completion_tokens: import_zod7.z.number().optional().describe("The successor to max_tokens"),
  frequency_penalty: import_zod7.z.number().optional(),
  presence_penalty: import_zod7.z.number().optional(),
  response_format: responseFormatSchema.nullish(),
  tool_choice: import_zod7.z.union([
    import_zod7.z.literal("auto").openapi({ title: "auto" }),
    import_zod7.z.literal("none").openapi({ title: "none" }),
    import_zod7.z.literal("required").openapi({ title: "required" }),
    import_zod7.z.object({
      type: import_zod7.z.literal("function"),
      function: import_zod7.z.object({ name: import_zod7.z.string() })
    }).openapi({ title: "function" })
  ]).optional(),
  function_call: import_zod7.z.union([
    import_zod7.z.literal("auto").openapi({ title: "auto" }),
    import_zod7.z.literal("none").openapi({ title: "none" }),
    import_zod7.z.object({
      name: import_zod7.z.string()
    }).openapi({ title: "function" })
  ]).optional(),
  n: import_zod7.z.number().optional(),
  stop: import_zod7.z.array(import_zod7.z.string()).optional(),
  reasoning_effort: import_zod7.z.enum(["low", "medium", "high"]).optional()
});
var anthropicModelParamsSchema = import_zod7.z.object({
  max_tokens: import_zod7.z.number(),
  temperature: import_zod7.z.number(),
  top_p: import_zod7.z.number().optional(),
  top_k: import_zod7.z.number().optional(),
  stop_sequences: import_zod7.z.array(import_zod7.z.string()).optional(),
  max_tokens_to_sample: import_zod7.z.number().optional().describe("This is a legacy parameter that should not be used.")
}).openapi("AntrhopicModelParams");
var googleModelParamsSchema = import_zod7.z.object({
  temperature: import_zod7.z.number().optional(),
  maxOutputTokens: import_zod7.z.number().optional(),
  topP: import_zod7.z.number().optional(),
  topK: import_zod7.z.number().optional()
}).openapi("GoogleModelParams");
var windowAIModelParamsSchema = import_zod7.z.object({
  temperature: import_zod7.z.number().optional(),
  topK: import_zod7.z.number().optional()
}).openapi("WindowAIModelParams");
var jsCompletionParamsSchema = import_zod7.z.object({}).openapi("JsCompletionParams");
var modelParamsSchema = import_zod7.z.union([
  braintrustModelParamsSchema.merge(openAIModelParamsSchema).passthrough().openapi({ title: "OpenAIModelParams" }),
  braintrustModelParamsSchema.merge(anthropicModelParamsSchema).passthrough().openapi({ title: "AnthropicModelParams" }),
  braintrustModelParamsSchema.merge(googleModelParamsSchema).passthrough().openapi({ title: "GoogleModelParams" }),
  braintrustModelParamsSchema.merge(windowAIModelParamsSchema).passthrough().openapi({ title: "WindowAIModelParams" }),
  braintrustModelParamsSchema.merge(jsCompletionParamsSchema).passthrough().openapi({ title: "JsCompletionParams" })
]).openapi("ModelParams");
var _anyModelParamsSchema = openAIModelParamsSchema.merge(anthropicModelParamsSchema).merge(googleModelParamsSchema).merge(braintrustModelParamsSchema);
var promptOptionsSchema = import_zod7.z.object({
  model: import_zod7.z.string().optional(),
  params: modelParamsSchema.optional(),
  position: import_zod7.z.string().optional()
}).openapi("PromptOptions");
var promptParserSchema = import_zod7.z.object({
  type: import_zod7.z.literal("llm_classifier"),
  use_cot: import_zod7.z.boolean(),
  choice_scores: import_zod7.z.record(import_zod7.z.number().min(0).max(1))
});
var promptDataSchema = import_zod7.z.object({
  prompt: promptBlockDataSchema.nullish(),
  options: promptOptionsSchema.nullish(),
  // This should be a union once we support multiple parser types
  parser: promptParserSchema.nullish(),
  tool_functions: import_zod7.z.array(savedFunctionIdSchema).nullish(),
  origin: import_zod7.z.object({
    prompt_id: import_zod7.z.string().optional(),
    project_id: import_zod7.z.string().optional(),
    prompt_version: import_zod7.z.string().optional()
  }).nullish()
}).openapi("PromptData");
var strictPromptDataSchema = promptDataSchema.extend({
  parser: promptParserSchema.extend({
    choice_scores: promptParserSchema.shape.choice_scores.refine(
      (r) => Object.keys(r).length > 0,
      "choice_scores must be nonempty"
    )
  }).nullish()
});

// typespecs/view.ts
var import_zod8 = require("zod");
var viewTypeEnum = import_zod8.z.enum([
  "projects",
  "experiments",
  "experiment",
  "playgrounds",
  "playground",
  "datasets",
  "dataset",
  "prompts",
  "tools",
  "scorers",
  "logs"
]).describe("Type of table that the view corresponds to.");
var viewDataSearchSchema = import_zod8.z.object({
  filter: import_zod8.z.array(customTypes.unknown).nullish(),
  tag: import_zod8.z.array(customTypes.unknown).nullish(),
  match: import_zod8.z.array(customTypes.unknown).nullish(),
  sort: import_zod8.z.array(customTypes.unknown).nullish()
}).strip().openapi("ViewDataSearch");
var viewDataSchema = import_zod8.z.object({
  search: viewDataSearchSchema.nullish()
}).strip().openapi("ViewData");
var viewOptionsSchema = import_zod8.z.object({
  columnVisibility: import_zod8.z.record(import_zod8.z.boolean()).nullish(),
  columnOrder: import_zod8.z.array(import_zod8.z.string()).nullish(),
  columnSizing: import_zod8.z.record(import_zod8.z.number()).nullish(),
  grouping: import_zod8.z.string().nullish(),
  rowHeight: import_zod8.z.string().nullish(),
  layout: import_zod8.z.string().nullish()
}).strip().openapi("ViewOptions");

// typespecs/functions.ts
var import_zod_to_openapi7 = require("@asteasolutions/zod-to-openapi");
var import_zod10 = require("zod");

// typespecs/git_types.ts
var import_zod_to_openapi6 = require("@asteasolutions/zod-to-openapi");
var import_zod9 = require("zod");
(0, import_zod_to_openapi6.extendZodWithOpenApi)(import_zod9.z);
var repoInfoSchema = import_zod9.z.object({
  commit: import_zod9.z.string().nullish().describe("SHA of most recent commit"),
  branch: import_zod9.z.string().nullish().describe("Name of the branch the most recent commit belongs to"),
  tag: import_zod9.z.string().nullish().describe("Name of the tag on the most recent commit"),
  dirty: import_zod9.z.boolean().nullish().describe(
    "Whether or not the repo had uncommitted changes when snapshotted"
  ),
  author_name: import_zod9.z.string().nullish().describe("Name of the author of the most recent commit"),
  author_email: import_zod9.z.string().nullish().describe("Email of the author of the most recent commit"),
  commit_message: import_zod9.z.string().nullish().describe("Most recent commit message"),
  commit_time: import_zod9.z.string().nullish().describe("Time of the most recent commit"),
  git_diff: import_zod9.z.string().nullish().describe(
    "If the repo was dirty when run, this includes the diff between the current state of the repo and the most recent commit."
  )
}).describe(
  "Metadata about the state of the repo when the experiment was created"
).openapi("RepoInfo");
var gitFieldsSchema = repoInfoSchema.keyof();
var collectMetadataEnum = import_zod9.z.enum(["all", "none", "some"]);
var gitMetadataSettingsSchema = import_zod9.z.strictObject({
  collect: collectMetadataEnum,
  fields: import_zod9.z.array(gitFieldsSchema).optional()
});

// typespecs/functions.ts
(0, import_zod_to_openapi7.extendZodWithOpenApi)(import_zod10.z);
var validRuntimesEnum = import_zod10.z.enum(["node", "python"]);
var functionTypeEnum = import_zod10.z.enum(["llm", "scorer", "task", "tool"]);
var runtimeContextSchema = import_zod10.z.object({
  runtime: validRuntimesEnum,
  version: import_zod10.z.string()
});
var functionIdSchema = import_zod10.z.union([
  import_zod10.z.object({
    function_id: import_zod10.z.string().describe("The ID of the function"),
    version: import_zod10.z.string().optional().describe("The version of the function")
  }).describe("Function id").openapi({ title: "function_id" }),
  import_zod10.z.object({
    project_name: import_zod10.z.string().describe("The name of the project containing the function"),
    slug: import_zod10.z.string().describe("The slug of the function"),
    version: import_zod10.z.string().optional().describe("The version of the function")
  }).describe("Project name and slug").openapi({ title: "project_slug" }),
  import_zod10.z.object({
    global_function: import_zod10.z.string().describe(
      "The name of the global function. Currently, the global namespace includes the functions in autoevals"
    )
  }).describe("Global function name").openapi({ title: "global_function" }),
  import_zod10.z.object({
    prompt_session_id: import_zod10.z.string().describe("The ID of the prompt session"),
    prompt_session_function_id: import_zod10.z.string().describe("The ID of the function in the prompt session"),
    version: import_zod10.z.string().optional().describe("The version of the function")
  }).describe("Prompt session id").openapi({ title: "prompt_session_id" }),
  import_zod10.z.object({
    inline_context: runtimeContextSchema,
    code: import_zod10.z.string().describe("The inline code to execute"),
    name: import_zod10.z.string().nullish().describe("The name of the inline code function")
  }).describe("Inline code function").openapi({ title: "inline_code" }),
  import_zod10.z.object({
    inline_prompt: promptDataSchema,
    name: import_zod10.z.string().nullish().describe("The name of the inline prompt")
  }).describe("Inline prompt definition").openapi({ title: "inline_prompt" })
]).describe("Options for identifying a function").openapi("FunctionId");
var useFunctionSchema = functionIdSchema;
var streamingModeEnum = import_zod10.z.enum(["auto", "parallel"]);
var invokeParent = import_zod10.z.union([
  import_zod10.z.object({
    object_type: import_zod10.z.enum(["project_logs", "experiment", "playground_logs"]),
    object_id: import_zod10.z.string().describe("The id of the container object you are logging to"),
    row_ids: import_zod10.z.object({
      id: import_zod10.z.string().describe("The id of the row"),
      span_id: import_zod10.z.string().describe("The span_id of the row"),
      root_span_id: import_zod10.z.string().describe("The root_span_id of the row")
    }).nullish().describe("Identifiers for the row to to log a subspan under"),
    propagated_event: import_zod10.z.record(customTypes.unknown).nullish().describe(
      "Include these properties in every span created under this parent"
    )
  }).describe("Span parent properties").openapi({ title: "span_parent_struct" }),
  import_zod10.z.string().optional().describe(
    "The parent's span identifier, created by calling `.export()` on a span"
  )
]);
var invokeFunctionNonIdArgsSchema = import_zod10.z.object({
  input: customTypes.unknown.optional().describe(
    "Argument to the function, which can be any JSON serializable value"
  ),
  expected: customTypes.unknown.optional().describe("The expected output of the function"),
  metadata: import_zod10.z.record(import_zod10.z.string(), import_zod10.z.unknown()).nullish().describe("Any relevant metadata"),
  messages: import_zod10.z.array(chatCompletionMessageParamSchema).optional().describe(
    "If the function is an LLM, additional messages to pass along to it"
  ),
  parent: invokeParent.describe("Options for tracing the function call"),
  stream: import_zod10.z.boolean().nullish().describe(
    "Whether to stream the response. If true, results will be returned in the Braintrust SSE format."
  ),
  mode: streamingModeEnum.nullish().describe("The mode format of the returned value (defaults to 'auto')")
});
var invokeFunctionSchema = functionIdSchema.and(invokeFunctionNonIdArgsSchema).openapi("InvokeFunction");
var invokeApiSchema = invokeFunctionNonIdArgsSchema.merge(
  import_zod10.z.object({
    version: import_zod10.z.string().optional().describe("The version of the function")
  })
).describe("The request to invoke a function").openapi("InvokeApi");
var runEvalSchema = import_zod10.z.object({
  project_id: import_zod10.z.string().describe("Unique identifier for the project to run the eval in"),
  data: import_zod10.z.union([
    import_zod10.z.object({
      dataset_id: import_zod10.z.string(),
      _internal_btql: import_zod10.z.record(import_zod10.z.string(), import_zod10.z.unknown()).nullish()
    }).describe("Dataset id").openapi({ title: "dataset_id" }),
    import_zod10.z.object({
      project_name: import_zod10.z.string(),
      dataset_name: import_zod10.z.string(),
      _internal_btql: import_zod10.z.record(import_zod10.z.string(), import_zod10.z.unknown()).nullish()
    }).describe("Project and dataset name").openapi({ title: "project_dataset_name" }),
    import_zod10.z.object({
      data: import_zod10.z.array(import_zod10.z.unknown())
      // TODO: More specific schema for these?
    }).describe("Dataset rows").openapi({ title: "dataset_rows" })
  ]).describe("The dataset to use"),
  task: functionIdSchema.describe("The function to evaluate"),
  scores: import_zod10.z.array(functionIdSchema).describe("The functions to score the eval on"),
  experiment_name: import_zod10.z.string().optional().describe(
    "An optional name for the experiment created by this eval. If it conflicts with an existing experiment, it will be suffixed with a unique identifier."
  ),
  metadata: import_zod10.z.record(customTypes.unknown).optional().describe(
    "Optional experiment-level metadata to store about the evaluation. You can later use this to slice & dice across experiments."
  ),
  parent: invokeParent.describe("Options for tracing the evaluation"),
  stream: import_zod10.z.boolean().optional().describe(
    "Whether to stream the results of the eval. If true, the request will return two events: one to indicate the experiment has started, and another upon completion. If false, the request will return the evaluation's summary upon completion."
  ),
  trial_count: import_zod10.z.number().nullish().describe(
    "The number of times to run the evaluator per input. This is useful for evaluating applications that have non-deterministic behavior and gives you both a stronger aggregate measure and a sense of the variance in the results."
  ),
  is_public: import_zod10.z.boolean().nullish().describe("Whether the experiment should be public. Defaults to false."),
  timeout: import_zod10.z.number().nullish().describe(
    "The maximum duration, in milliseconds, to run the evaluation. Defaults to undefined, in which case there is no timeout."
  ),
  max_concurrency: import_zod10.z.number().nullish().describe(
    "The maximum number of tasks/scorers that will be run concurrently. Defaults to undefined, in which case there is no max concurrency."
  ),
  base_experiment_name: import_zod10.z.string().nullish().describe(
    "An optional experiment name to use as a base. If specified, the new experiment will be summarized and compared to this experiment."
  ),
  base_experiment_id: import_zod10.z.string().nullish().describe(
    "An optional experiment id to use as a base. If specified, the new experiment will be summarized and compared to this experiment."
  ),
  git_metadata_settings: gitMetadataSettingsSchema.nullish().describe(
    "Optional settings for collecting git metadata. By default, will collect all git metadata fields allowed in org-level settings."
  ),
  repo_info: repoInfoSchema.nullish().describe(
    "Optionally explicitly specify the git metadata for this experiment. This takes precedence over `gitMetadataSettings` if specified."
  )
}).openapi("RunEval");
var baseSSEEventSchema = import_zod10.z.object({
  id: import_zod10.z.string().optional(),
  data: import_zod10.z.string()
});
var sseTextEventSchema = baseSSEEventSchema.merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("text_delta")
  })
);
var sseDataEventSchema = baseSSEEventSchema.merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("json_delta")
  })
);
var sseErrorEventSchema = baseSSEEventSchema.merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("error")
  })
);
var sseProgressEventSchema = baseSSEEventSchema.merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("progress")
  })
);
var sseConsoleEventSchema = baseSSEEventSchema.merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("console")
  })
);
var sseStartEventSchema = baseSSEEventSchema.omit({ data: true }).merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("start"),
    data: import_zod10.z.literal("")
  })
);
var sseDoneEventSchema = baseSSEEventSchema.omit({ data: true }).merge(
  import_zod10.z.object({
    event: import_zod10.z.literal("done"),
    data: import_zod10.z.literal("")
  })
);
var functionObjectTypeEnum = import_zod10.z.enum(["prompt", "tool", "scorer", "task"]).openapi("FunctionObjectType");
var functionFormatEnum = import_zod10.z.enum(["llm", "code", "global"]).openapi("FunctionFormat");
var functionOutputTypeEnum = import_zod10.z.enum(["completion", "score", "any"]).openapi("FunctionOutputType");
var sseProgressEventDataSchema = import_zod10.z.object({
  id: import_zod10.z.string().describe("The id of the span this event is for"),
  object_type: functionObjectTypeEnum,
  origin: objectReferenceSchema.nullish().describe("The origin of the event"),
  format: functionFormatEnum,
  output_type: functionOutputTypeEnum,
  name: import_zod10.z.string(),
  event: import_zod10.z.enum([
    "text_delta",
    "json_delta",
    "error",
    "console",
    "start",
    "done",
    "progress"
  ]),
  data: import_zod10.z.string()
  // This is the text_delta or json_delta
}).openapi("SSEProgressEventData");
var sseConsoleEventDataSchema = import_zod10.z.object({
  stream: import_zod10.z.enum(["stderr", "stdout"]),
  message: import_zod10.z.string()
});
var callEventSchema = import_zod10.z.union([
  sseTextEventSchema.openapi({ title: "text_delta" }),
  sseDataEventSchema.openapi({ title: "json_delta" }),
  sseProgressEventSchema.openapi({ title: "progress" }),
  sseErrorEventSchema.openapi({ title: "error" }),
  sseConsoleEventSchema.openapi({ title: "console" }),
  sseStartEventSchema.openapi({ title: "start" }),
  sseDoneEventSchema.openapi({ title: "done" })
]).openapi("CallEvent");
var scoreSchema = import_zod10.z.union([
  import_zod10.z.object({
    name: import_zod10.z.string(),
    score: import_zod10.z.number().min(0).max(1).nullable().default(null),
    // Sometimes we get an empty value over the wire
    metadata: import_zod10.z.record(customTypes.unknown).optional().transform((data) => data != null ? data : void 0)
  }),
  import_zod10.z.number().min(0).max(1),
  import_zod10.z.boolean().transform((b) => b ? 1 : 0),
  import_zod10.z.null()
]).openapi("ScorerScore");
var ifExistsEnum = import_zod10.z.enum(["error", "ignore", "replace"]);
var DEFAULT_IF_EXISTS = "error";
var toolFunctionDefinitionSchema = import_zod10.z.object({
  type: import_zod10.z.literal("function"),
  function: import_zod10.z.object({
    name: import_zod10.z.string(),
    description: import_zod10.z.string().optional(),
    parameters: import_zod10.z.record(import_zod10.z.unknown()).optional(),
    strict: import_zod10.z.boolean().optional()
  })
});

// typespecs/app_types.ts
(0, import_zod_to_openapi8.extendZodWithOpenApi)(import_zod11.z);
function generateBaseTableSchema(objectName, opts) {
  let nameDescription = `Name of the ${objectName}`;
  if (opts == null ? void 0 : opts.uniqueName) {
    nameDescription += `. Within a project, ${objectName} names are unique`;
  }
  return import_zod11.z.object({
    id: import_zod11.z.string().uuid().describe(`Unique identifier for the ${objectName}`),
    project_id: import_zod11.z.string().uuid().describe(
      `Unique identifier for the project that the ${objectName} belongs under`
    ),
    name: import_zod11.z.string().describe(nameDescription),
    description: import_zod11.z.string().nullish().describe(`Textual description of the ${objectName}`),
    created: datetimeStringSchema.nullish().describe(`Date of ${objectName} creation`),
    deleted_at: datetimeStringSchema.nullish().describe(
      `Date of ${objectName} deletion, or null if the ${objectName} is still active`
    ),
    updated_at: datetimeStringSchema.nullish().describe(`Date of last ${objectName} update`),
    user_id: import_zod11.z.string().uuid().nullish().describe(`Identifies the user who created the ${objectName}`),
    metadata: import_zod11.z.record(customTypes.unknown).nullish().describe(`User-controlled metadata about the ${objectName}`)
  });
}
var aclObjectTypeEnum = import_zod11.z.enum([
  "organization",
  "project",
  "experiment",
  "dataset",
  "prompt",
  "prompt_session",
  "group",
  "role",
  "org_member",
  "project_log",
  "org_project"
]).describe("The object type that the ACL applies to").openapi("AclObjectType");
var userBaseSchema = generateBaseTableSchema("user");
var userSchema = import_zod11.z.object({
  id: userBaseSchema.shape.id,
  given_name: import_zod11.z.string().nullish().describe("Given name of the user"),
  family_name: import_zod11.z.string().nullish().describe("Family name of the user"),
  email: import_zod11.z.string().nullish().describe("The user's email"),
  avatar_url: import_zod11.z.string().nullish().describe("URL of the user's Avatar image"),
  created: userBaseSchema.shape.created
}).openapi("User");
var organizationBaseSchema = generateBaseTableSchema("organization");
var organizationSchema = import_zod11.z.object({
  id: organizationBaseSchema.shape.id,
  name: organizationBaseSchema.shape.name,
  api_url: import_zod11.z.string().nullish(),
  is_universal_api: import_zod11.z.boolean().nullish(),
  proxy_url: import_zod11.z.string().nullish(),
  realtime_url: import_zod11.z.string().nullish(),
  created: organizationBaseSchema.shape.created
}).openapi("Organization");
var maxOverWindowSchema = import_zod11.z.object({
  window_size_days: import_zod11.z.number().int().positive(),
  max_value: import_zod11.z.number().nonnegative()
}).openapi("MaxOverWindow");
var resourcesSchema = import_zod11.z.object({
  org_id: organizationSchema.shape.id,
  forbid_toggle_experiment_public_to_private: import_zod11.z.boolean().nullish(),
  num_private_experiment_row_actions: maxOverWindowSchema.nullish(),
  forbid_insert_datasets: import_zod11.z.boolean().nullish(),
  forbid_insert_prompt_sessions: import_zod11.z.boolean().nullish(),
  forbid_access_sql_explorer: import_zod11.z.boolean().nullish(),
  num_production_log_row_actions: maxOverWindowSchema.nullish(),
  num_dataset_row_actions: maxOverWindowSchema.nullish()
}).openapi("Resources");
var memberSchema = import_zod11.z.object({
  org_id: organizationSchema.shape.id,
  user_id: userSchema.shape.id
}).openapi("Member");
var aiSecretBaseSchema = generateBaseTableSchema("AI secret");
var aiSecretSchema = import_zod11.z.object({
  id: aiSecretBaseSchema.shape.id,
  created: aiSecretBaseSchema.shape.created,
  updated_at: aiSecretBaseSchema.shape.updated_at,
  org_id: organizationSchema.shape.id,
  name: aiSecretBaseSchema.shape.name,
  type: import_zod11.z.string().nullish(),
  metadata: import_zod11.z.record(customTypes.unknown).nullish(),
  preview_secret: import_zod11.z.string().nullish()
}).openapi("AISecret");
var envVarObjectTypeEnum = import_zod11.z.enum(["organization", "project", "function"]).describe("The type of the object the environment variable is scoped for");
var envVarBaseSchema = generateBaseTableSchema("environment variable");
var envVarSchema = import_zod11.z.object({
  id: envVarBaseSchema.shape.id,
  object_type: envVarObjectTypeEnum,
  object_id: import_zod11.z.string().uuid().describe("The id of the object the environment variable is scoped for"),
  name: import_zod11.z.string().describe("The name of the environment variable"),
  created: envVarBaseSchema.shape.created,
  used: datetimeStringSchema.nullish().describe(`Date the environment variable was last used`)
}).openapi("EnvVar");
var customColumnBaseSchema = generateBaseTableSchema("custom columns");
var customColumnSchema = import_zod11.z.object({
  id: customColumnBaseSchema.shape.id,
  object_type: aclObjectTypeEnum,
  object_id: import_zod11.z.string().uuid().describe("The id of the object the custom column is scoped for"),
  subtype: aclObjectTypeEnum.nullable(),
  name: import_zod11.z.string().describe("The name of the custom column"),
  expr: import_zod11.z.string().describe(
    "The expression used to extract the value for the custom column"
  ),
  created: customColumnBaseSchema.shape.created
}).openapi("CustomColumn");
var apiKeyBaseSchema = generateBaseTableSchema("api key");
var apiKeySchema = import_zod11.z.object({
  id: apiKeyBaseSchema.shape.id,
  created: apiKeyBaseSchema.shape.created,
  name: apiKeyBaseSchema.shape.name,
  preview_name: import_zod11.z.string(),
  user_id: userSchema.shape.id.nullish(),
  org_id: organizationSchema.shape.id.nullish()
}).openapi("ApiKey");
var spanFieldOrderItem = import_zod11.z.object({
  object_type: import_zod11.z.string(),
  column_id: import_zod11.z.string(),
  position: import_zod11.z.string(),
  layout: import_zod11.z.literal("full").or(import_zod11.z.literal("two_column")).nullish()
});
var projectSettingsSchema = import_zod11.z.object({
  comparison_key: import_zod11.z.string().nullish().describe("The key used to join two experiments (defaults to `input`)"),
  baseline_experiment_id: import_zod11.z.string().uuid().nullish().describe(
    "The id of the experiment to use as the default baseline for comparisons"
  ),
  spanFieldOrder: import_zod11.z.array(spanFieldOrderItem).nullish().describe("The order of the fields to display in the trace view")
}).openapi("ProjectSettings");
var projectBaseSchema = generateBaseTableSchema("project");
var projectSchema = import_zod11.z.object({
  id: projectBaseSchema.shape.id,
  org_id: import_zod11.z.string().uuid().describe(
    "Unique id for the organization that the project belongs under"
  ),
  name: projectBaseSchema.shape.name,
  created: projectBaseSchema.shape.created,
  deleted_at: projectBaseSchema.shape.deleted_at,
  user_id: projectBaseSchema.shape.user_id,
  settings: projectSettingsSchema.nullish()
}).openapi("Project");
var datasetBaseSchema = generateBaseTableSchema("dataset", {
  uniqueName: true
});
var datasetSchema = import_zod11.z.object({
  id: datasetBaseSchema.shape.id,
  project_id: datasetBaseSchema.shape.project_id,
  name: datasetBaseSchema.shape.name,
  description: datasetBaseSchema.shape.description,
  created: datasetBaseSchema.shape.created,
  deleted_at: datasetBaseSchema.shape.deleted_at,
  user_id: datasetBaseSchema.shape.user_id,
  metadata: datasetBaseSchema.shape.metadata
}).openapi("Dataset");
var promptLogIdLiteralSchema = import_zod11.z.literal("p").describe("A literal 'p' which identifies the object as a project prompt");
var playgroundLogsLogIdLiteralSchema = import_zod11.z.literal("x").describe("A literal 'x' which identifies the object as a playground log");
var promptBaseSchema = generateBaseTableSchema("prompt");
var promptSchemaObject = import_zod11.z.object({
  id: promptBaseSchema.shape.id,
  // This has to be copy/pasted because zod blows up when there are circular dependencies
  _xact_id: import_zod11.z.string().describe(
    `The transaction id of an event is unique to the network operation that processed the event insertion. Transaction ids are monotonically increasing over time and can be used to retrieve a versioned snapshot of the prompt (see the \`version\` parameter)`
  ),
  project_id: promptBaseSchema.shape.project_id,
  log_id: promptLogIdLiteralSchema,
  org_id: organizationSchema.shape.id,
  name: promptBaseSchema.shape.name,
  slug: import_zod11.z.string().describe("Unique identifier for the prompt"),
  description: promptBaseSchema.shape.description,
  created: promptBaseSchema.shape.created,
  prompt_data: promptDataSchema.nullish().describe("The prompt, model, and its parameters"),
  tags: import_zod11.z.array(import_zod11.z.string()).nullish().describe("A list of tags for the prompt"),
  metadata: promptBaseSchema.shape.metadata,
  // An empty (unspecified) function_type is equivalent to "task".
  function_type: functionTypeEnum.nullish()
});
var promptSchema = promptSchemaObject.openapi("Prompt");
var codeBundleSchema = import_zod11.z.object({
  runtime_context: runtimeContextSchema,
  location: import_zod11.z.union([
    import_zod11.z.object({
      type: import_zod11.z.literal("experiment"),
      eval_name: import_zod11.z.string(),
      position: import_zod11.z.union([
        import_zod11.z.object({ type: import_zod11.z.literal("task") }),
        import_zod11.z.object({
          type: import_zod11.z.literal("scorer"),
          index: import_zod11.z.number().int().nonnegative()
        }).openapi({ title: "scorer" })
      ])
    }).openapi({ title: "experiment" }),
    import_zod11.z.object({
      type: import_zod11.z.literal("function"),
      index: import_zod11.z.number().int().nonnegative()
    }).openapi({ title: "function" })
  ]),
  bundle_id: import_zod11.z.string(),
  preview: import_zod11.z.string().nullish().describe("A preview of the code")
}).openapi("CodeBundle");
var functionDataSchema = import_zod11.z.union([
  import_zod11.z.object({
    type: import_zod11.z.literal("prompt")
    // For backwards compatibility reasons, the prompt definition is hoisted out and stored
    // in the outer object
  }).openapi({ title: "prompt" }),
  import_zod11.z.object({
    type: import_zod11.z.literal("code"),
    data: import_zod11.z.union([
      import_zod11.z.object({
        type: import_zod11.z.literal("bundle")
      }).and(codeBundleSchema).openapi({ title: "bundle" }),
      import_zod11.z.object({
        type: import_zod11.z.literal("inline"),
        runtime_context: runtimeContextSchema,
        code: import_zod11.z.string()
      }).openapi({ title: "inline" })
    ])
  }).openapi({ title: "code" }),
  import_zod11.z.object({
    type: import_zod11.z.literal("global"),
    name: import_zod11.z.string()
  }).openapi({ title: "global" })
]).openapi("FunctionData");
var functionSchema2 = promptSchemaObject.merge(
  import_zod11.z.object({
    function_data: functionDataSchema,
    origin: import_zod11.z.object({
      object_type: aclObjectTypeEnum,
      object_id: import_zod11.z.string().uuid().describe("Id of the object the function is originating from"),
      internal: import_zod11.z.boolean().nullish().describe(
        "The function exists for internal purposes and should not be displayed in the list of functions."
      )
    }).nullish(),
    function_schema: import_zod11.z.object({
      parameters: customTypes.unknown,
      returns: customTypes.unknown.optional()
    }).nullish().describe("JSON schema for the function's parameters and return type")
  })
).openapi("Function");
var experimentBaseSchema = generateBaseTableSchema("experiment", {
  uniqueName: true
});
var experimentSchema = import_zod11.z.object({
  id: experimentBaseSchema.shape.id,
  project_id: experimentBaseSchema.shape.project_id,
  name: experimentBaseSchema.shape.name,
  description: experimentBaseSchema.shape.description,
  created: experimentBaseSchema.shape.created,
  repo_info: repoInfoSchema.nullish(),
  commit: import_zod11.z.string().nullish().describe("Commit, taken directly from `repo_info.commit`"),
  base_exp_id: import_zod11.z.string().uuid().nullish().describe(
    "Id of default base experiment to compare against when viewing this experiment"
  ),
  deleted_at: experimentBaseSchema.shape.deleted_at,
  dataset_id: import_zod11.z.string().uuid().nullish().describe(
    "Identifier of the linked dataset, or null if the experiment is not linked to a dataset"
  ),
  dataset_version: import_zod11.z.string().nullish().describe(
    "Version number of the linked dataset the experiment was run against. This can be used to reproduce the experiment after the dataset has been modified."
  ),
  public: import_zod11.z.boolean().describe(
    "Whether or not the experiment is public. Public experiments can be viewed by anybody inside or outside the organization"
  ),
  user_id: experimentBaseSchema.shape.user_id,
  metadata: experimentBaseSchema.shape.metadata
}).openapi("Experiment");
var promptSessionBaseSchema = generateBaseTableSchema("promptSession", {
  uniqueName: true
});
var promptSessionSchema = import_zod11.z.object({
  id: promptSessionBaseSchema.shape.id,
  name: promptSessionBaseSchema.shape.name,
  description: promptSessionBaseSchema.shape.description,
  created: promptSessionBaseSchema.shape.created,
  deleted_at: promptSessionBaseSchema.shape.deleted_at,
  user_id: promptSessionBaseSchema.shape.user_id,
  project_id: promptSessionBaseSchema.shape.project_id,
  org_id: organizationSchema.shape.id.nullish().describe(
    "This field is deprecated and will be removed in a future revision"
  )
}).openapi("PromptSession");
var permissionEnum = import_zod11.z.enum([
  "create",
  "read",
  "update",
  "delete",
  "create_acls",
  "read_acls",
  "update_acls",
  "delete_acls"
]).describe(
  [
    "Each permission permits a certain type of operation on an object in the system",
    "Permissions can be assigned to to objects on an individual basis, or grouped into roles"
  ].join("\n\n")
).openapi("Permission");
var roleBaseSchema = generateBaseTableSchema("role");
var roleSchema = import_zod11.z.object({
  id: roleBaseSchema.shape.id,
  org_id: import_zod11.z.string().uuid().nullish().describe(
    [
      "Unique id for the organization that the role belongs under",
      "A null org_id indicates a system role, which may be assigned to anybody and inherited by any other role, but cannot be edited.",
      "It is forbidden to change the org after creating a role"
    ].join("\n\n")
  ),
  user_id: roleBaseSchema.shape.user_id,
  created: roleBaseSchema.shape.created,
  name: roleBaseSchema.shape.name,
  description: roleBaseSchema.shape.description,
  deleted_at: roleBaseSchema.shape.deleted_at,
  member_permissions: import_zod11.z.array(
    import_zod11.z.object({
      permission: permissionEnum,
      restrict_object_type: aclObjectTypeEnum.nullish()
    })
  ).nullish().describe(
    "(permission, restrict_object_type) tuples which belong to this role"
  ),
  member_roles: import_zod11.z.array(import_zod11.z.string().uuid()).nullish().describe(
    [
      "Ids of the roles this role inherits from",
      "An inheriting role has all the permissions contained in its member roles, as well as all of their inherited permissions"
    ].join("\n\n")
  )
}).describe(
  [
    "A role is a collection of permissions which can be granted as part of an ACL",
    "Roles can consist of individual permissions, as well as a set of roles they inherit from"
  ].join("\n\n")
).openapi("Role");
var groupBaseSchema = generateBaseTableSchema("group");
var groupSchema = import_zod11.z.object({
  id: groupBaseSchema.shape.id,
  org_id: import_zod11.z.string().uuid().describe(
    [
      "Unique id for the organization that the group belongs under",
      "It is forbidden to change the org after creating a group"
    ].join("\n\n")
  ),
  user_id: groupBaseSchema.shape.user_id,
  created: groupBaseSchema.shape.created,
  name: groupBaseSchema.shape.name,
  description: groupBaseSchema.shape.description,
  deleted_at: groupBaseSchema.shape.deleted_at,
  member_users: import_zod11.z.array(import_zod11.z.string().uuid()).nullish().describe("Ids of users which belong to this group"),
  member_groups: import_zod11.z.array(import_zod11.z.string().uuid()).nullish().describe(
    [
      "Ids of the groups this group inherits from",
      "An inheriting group has all the users contained in its member groups, as well as all of their inherited users"
    ].join("\n\n")
  )
}).describe(
  [
    "A group is a collection of users which can be assigned an ACL",
    "Groups can consist of individual users, as well as a set of groups they inherit from"
  ].join("\n\n")
).openapi("Group");
var projectScoreTypeEnum = import_zod11.z.enum([
  "slider",
  "categorical",
  "weighted",
  "minimum",
  "maximum",
  "online",
  "free-form"
]).describe("The type of the configured score").openapi("ProjectScoreType");
var projectScoreCategory = import_zod11.z.object({
  name: import_zod11.z.string().describe("Name of the category"),
  value: import_zod11.z.number().describe(
    "Numerical value of the category. Must be between 0 and 1, inclusive"
  )
}).describe("For categorical-type project scores, defines a single category").openapi("ProjectScoreCategory");
var onlineScoreConfigSchema = import_zod11.z.object({
  sampling_rate: import_zod11.z.number().min(0).max(1).describe("The sampling rate for online scoring"),
  scorers: import_zod11.z.array(savedFunctionIdSchema).describe("The list of scorers to use for online scoring"),
  apply_to_root_span: import_zod11.z.boolean().nullish().describe(
    "Whether to trigger online scoring on the root span of each trace"
  ),
  apply_to_span_names: import_zod11.z.string().array().nullish().describe("Trigger online scoring on any spans with a name in this list")
}).refine((val) => {
  var _a;
  return val.apply_to_root_span || ((_a = val.apply_to_span_names) == null ? void 0 : _a.length);
}, {
  message: "Online scoring rule does not apply to any rows"
}).openapi("OnlineScoreConfig");
var projectScoreBaseSchema = generateBaseTableSchema("project score");
var projectScoreSchema = import_zod11.z.object({
  id: projectScoreBaseSchema.shape.id,
  project_id: projectScoreBaseSchema.shape.project_id,
  user_id: projectScoreBaseSchema.shape.user_id.unwrap().unwrap(),
  created: projectScoreBaseSchema.shape.created,
  name: projectScoreBaseSchema.shape.name,
  description: projectScoreBaseSchema.shape.description,
  score_type: projectScoreTypeEnum,
  categories: import_zod11.z.union([
    projectScoreCategory.array().describe(
      "For categorical-type project scores, the list of all categories"
    ).openapi({ title: "categorical" }),
    import_zod11.z.record(import_zod11.z.number()).describe(
      "For weighted-type project scores, the weights of each score"
    ).openapi({ title: "weighted" }),
    import_zod11.z.array(import_zod11.z.string()).describe(
      "For minimum-type project scores, the list of included scores"
    ).openapi({ title: "minimum" })
  ]).nullish().openapi("ProjectScoreCategories"),
  config: import_zod11.z.object({
    multi_select: import_zod11.z.boolean().nullish(),
    destination: import_zod11.z.string().nullish(),
    online: onlineScoreConfigSchema.nullish()
  }).nullish().openapi("ProjectScoreConfig"),
  position: import_zod11.z.string().nullish().describe(
    "An optional LexoRank-based string that sets the sort position for the score in the UI"
  )
}).describe(
  "A project score is a user-configured score, which can be manually-labeled through the UI"
).openapi("ProjectScore");
var projectTagBaseSchema = generateBaseTableSchema("project tag");
var projectTagSchema = import_zod11.z.object({
  id: projectTagBaseSchema.shape.id,
  project_id: projectTagBaseSchema.shape.project_id,
  user_id: projectTagBaseSchema.shape.user_id.unwrap().unwrap(),
  created: projectTagBaseSchema.shape.created,
  name: projectTagBaseSchema.shape.name,
  description: projectTagBaseSchema.shape.description,
  color: import_zod11.z.string().nullish().describe("Color of the tag for the UI")
}).describe(
  "A project tag is a user-configured tag for tracking and filtering your experiments, logs, and other data"
).openapi("ProjectTag");
var spanIframeBaseSchema = generateBaseTableSchema("span iframe");
var spanIframeSchema = import_zod11.z.object({
  id: spanIframeBaseSchema.shape.id,
  project_id: spanIframeBaseSchema.shape.project_id,
  user_id: spanIframeBaseSchema.shape.user_id,
  created: spanIframeBaseSchema.shape.created,
  deleted_at: spanIframeBaseSchema.shape.deleted_at,
  name: spanIframeBaseSchema.shape.name,
  description: spanIframeBaseSchema.shape.description,
  url: import_zod11.z.string().describe("URL to embed the project viewer in an iframe"),
  post_message: import_zod11.z.boolean().nullish().describe(
    "Whether to post messages to the iframe containing the span's data. This is useful when you want to render more data than fits in the URL."
  )
}).openapi("SpanIFrame");
var viewBaseSchema = generateBaseTableSchema("view");
var viewSchema = import_zod11.z.object({
  id: viewBaseSchema.shape.id,
  object_type: aclObjectTypeEnum,
  object_id: import_zod11.z.string().uuid().describe("The id of the object the view applies to"),
  view_type: viewTypeEnum,
  name: viewBaseSchema.shape.name,
  created: viewBaseSchema.shape.created,
  view_data: viewDataSchema.nullish().describe("The view definition"),
  options: viewOptionsSchema.nullish().describe("Options for the view in the app"),
  user_id: viewBaseSchema.shape.user_id,
  deleted_at: roleBaseSchema.shape.deleted_at
}).openapi("View");
var aclBaseSchema = generateBaseTableSchema("acl");
var aclSchema = import_zod11.z.object({
  id: aclBaseSchema.shape.id,
  object_type: aclObjectTypeEnum,
  object_id: import_zod11.z.string().uuid().describe("The id of the object the ACL applies to"),
  user_id: import_zod11.z.string().uuid().nullish().describe(
    "Id of the user the ACL applies to. Exactly one of `user_id` and `group_id` will be provided"
  ),
  group_id: import_zod11.z.string().uuid().nullish().describe(
    "Id of the group the ACL applies to. Exactly one of `user_id` and `group_id` will be provided"
  ),
  permission: permissionEnum.nullish().describe(
    "Permission the ACL grants. Exactly one of `permission` and `role_id` will be provided"
  ),
  restrict_object_type: aclObjectTypeEnum.nullish().describe(
    "When setting a permission directly, optionally restricts the permission grant to just the specified object type. Cannot be set alongside a `role_id`."
  ),
  role_id: import_zod11.z.string().uuid().nullish().describe(
    "Id of the role the ACL grants. Exactly one of `permission` and `role_id` will be provided"
  ),
  _object_org_id: import_zod11.z.string().uuid().describe("The organization the ACL's referred object belongs to"),
  created: aclBaseSchema.shape.created
}).describe(
  [
    "An ACL grants a certain permission or role to a certain user or group on an object.",
    "ACLs are inherited across the object hierarchy. So for example, if a user has read permissions on a project, they will also have read permissions on any experiment, dataset, etc. created within that project.",
    "To restrict a grant to a particular sub-object, you may specify `restrict_object_type` in the ACL, as part of a direct permission grant or as part of a role."
  ].join("\n\n")
).openapi("Acl");
var appLimitParamSchema = import_zod11.z.coerce.number().int().nonnegative().describe("Limit the number of objects to return").openapi("AppLimit");
function generateBaseTableOpSchema(objectName) {
  return import_zod11.z.object({
    org_name: import_zod11.z.string().nullish().describe(
      `For nearly all users, this parameter should be unnecessary. But in the rare case that your API key belongs to multiple organizations, you may specify the name of the organization the ${objectName} belongs in.`
    )
  });
}
var startingAfterSchema = import_zod11.z.string().uuid().describe(
  [
    "Pagination cursor id.",
    "For example, if the final item in the last page you fetched had an id of `foo`, pass `starting_after=foo` to fetch the next page. Note: you may only pass one of `starting_after` and `ending_before`"
  ].join("\n\n")
).openapi("StartingAfter");
var endingBeforeSchema = import_zod11.z.string().uuid().describe(
  [
    "Pagination cursor id.",
    "For example, if the initial item in the last page you fetched had an id of `foo`, pass `ending_before=foo` to fetch the previous page. Note: you may only pass one of `starting_after` and `ending_before`"
  ].join("\n\n")
).openapi("EndingBefore");
function makeObjectIdsFilterSchema(objectName) {
  const item = import_zod11.z.string().uuid();
  return import_zod11.z.union([item, item.array()]).describe(
    `Filter search results to a particular set of ${objectName} IDs. To specify a list of IDs, include the query param multiple times`
  ).openapi(`${objectName}IdsFilter`);
}
function makeNonempty(s) {
  var _a;
  return ((_a = s.minLength) != null ? _a : 0) > 0 ? s : s.min(1);
}
var createProjectBaseSchema = generateBaseTableOpSchema("project");
var createProjectSchema = import_zod11.z.object({
  name: makeNonempty(projectSchema.shape.name),
  org_name: createProjectBaseSchema.shape.org_name
}).openapi("CreateProject");
var patchProjectSchema = import_zod11.z.object({
  name: projectSchema.shape.name.nullish(),
  settings: projectSchema.shape.settings.describe(
    "Project settings. Patch operations replace all settings, so make sure you include all settings you want to keep."
  ).nullish()
}).openapi("PatchProject");
var createExperimentSchema = import_zod11.z.object({
  project_id: experimentSchema.shape.project_id,
  name: makeNonempty(experimentSchema.shape.name).nullish(),
  description: experimentSchema.shape.description,
  repo_info: experimentSchema.shape.repo_info,
  base_exp_id: experimentSchema.shape.base_exp_id,
  dataset_id: experimentSchema.shape.dataset_id,
  dataset_version: experimentSchema.shape.dataset_version,
  public: experimentSchema.shape.public.nullish(),
  metadata: experimentSchema.shape.metadata,
  ensure_new: import_zod11.z.boolean().nullish().describe(
    "Normally, creating an experiment with the same name as an existing experiment will return the existing one un-modified. But if `ensure_new` is true, registration will generate a new experiment with a unique name in case of a conflict."
  )
}).openapi("CreateExperiment");
var patchExperimentSchema = createExperimentSchema.omit({ project_id: true, ensure_new: true }).extend({ name: experimentSchema.shape.name.nullish() }).openapi("PatchExperiment");
var createDatasetSchema = import_zod11.z.object({
  project_id: datasetSchema.shape.project_id,
  name: makeNonempty(datasetSchema.shape.name),
  description: datasetSchema.shape.description,
  metadata: datasetSchema.shape.metadata
}).openapi("CreateDataset");
var patchDatasetSchema = import_zod11.z.object({
  name: datasetSchema.shape.name.nullish(),
  description: datasetSchema.shape.description,
  metadata: datasetSchema.shape.metadata
}).openapi("PatchDataset");
var createPromptSchema = promptSchema.omit({
  id: true,
  _xact_id: true,
  org_id: true,
  log_id: true,
  created: true,
  metadata: true
}).extend({
  name: makeNonempty(promptSchema.shape.name),
  slug: makeNonempty(promptSchema.shape.slug)
}).openapi("CreatePrompt");
var createFunctionSchema = functionSchema2.omit({
  id: true,
  _xact_id: true,
  org_id: true,
  log_id: true,
  created: true,
  metadata: true
}).extend({
  name: makeNonempty(promptSchema.shape.name),
  slug: makeNonempty(promptSchema.shape.slug)
}).openapi("CreateFunction");
var patchPromptSchema = import_zod11.z.object({
  name: promptSchema.shape.name.nullish(),
  slug: promptSchema.shape.slug.nullish(),
  description: promptSchema.shape.description.nullish(),
  prompt_data: promptSchema.shape.prompt_data.nullish(),
  tags: promptSchema.shape.tags.nullish()
}).openapi("PatchPrompt");
var patchFunctionSchema = import_zod11.z.object({
  name: functionSchema2.shape.name.nullish(),
  description: functionSchema2.shape.description.nullish(),
  prompt_data: functionSchema2.shape.prompt_data.nullish(),
  function_data: functionSchema2.shape.function_data.nullish().openapi("FunctionDataNullish"),
  tags: functionSchema2.shape.tags.nullish()
}).openapi("PatchFunction");
var createRoleBaseSchema = generateBaseTableOpSchema("role");
var createRoleSchema = import_zod11.z.object({
  name: makeNonempty(roleSchema.shape.name),
  description: roleSchema.shape.description,
  member_permissions: roleSchema.shape.member_permissions,
  member_roles: roleSchema.shape.member_roles,
  org_name: createRoleBaseSchema.shape.org_name
}).openapi("CreateRole");
var patchRoleSchema = createRoleSchema.omit({
  name: true,
  org_name: true,
  member_permissions: true,
  member_roles: true
}).merge(
  import_zod11.z.object({
    name: createRoleSchema.shape.name.nullish(),
    add_member_permissions: roleSchema.shape.member_permissions.nullish().describe("A list of permissions to add to the role"),
    remove_member_permissions: roleSchema.shape.member_permissions.nullish().describe("A list of permissions to remove from the role"),
    add_member_roles: roleSchema.shape.member_roles.nullish().describe(
      "A list of role IDs to add to the role's inheriting-from set"
    ),
    remove_member_roles: roleSchema.shape.member_roles.nullish().describe(
      "A list of role IDs to remove from the role's inheriting-from set"
    )
  })
).openapi("PatchRole");
var createGroupBaseSchema = generateBaseTableOpSchema("group");
var createGroupSchema = import_zod11.z.object({
  name: makeNonempty(groupSchema.shape.name),
  description: groupSchema.shape.description,
  member_users: groupSchema.shape.member_users,
  member_groups: groupSchema.shape.member_groups,
  org_name: createGroupBaseSchema.shape.org_name
}).openapi("CreateGroup");
var patchGroupSchema = createGroupSchema.omit({ name: true, org_name: true, member_users: true, member_groups: true }).merge(
  import_zod11.z.object({
    name: createGroupSchema.shape.name.nullish(),
    add_member_users: groupSchema.shape.member_users.nullish().describe("A list of user IDs to add to the group"),
    remove_member_users: groupSchema.shape.member_users.nullish().describe("A list of user IDs to remove from the group"),
    add_member_groups: groupSchema.shape.member_groups.nullish().describe(
      "A list of group IDs to add to the group's inheriting-from set"
    ),
    remove_member_groups: groupSchema.shape.member_groups.nullish().describe(
      "A list of group IDs to remove from the group's inheriting-from set"
    )
  })
).openapi("PatchGroup");
var aclItemSchema = aclSchema.omit({
  id: true,
  created: true,
  _object_org_id: true
}).openapi("AclItem");
var aclBatchUpdateRequestSchema = import_zod11.z.object({
  add_acls: aclItemSchema.array().nullish(),
  remove_acls: aclItemSchema.array().nullish()
}).openapi("AclBatchUpdateRequest");
var aclBatchUpdateResponseSchema = import_zod11.z.object({
  added_acls: aclSchema.array(),
  removed_acls: aclSchema.array()
}).openapi("AclBatchUpdateResponse");
var createProjectScoreSchema = projectScoreSchema.pick({
  project_id: true,
  name: true,
  description: true,
  score_type: true,
  categories: true,
  config: true
}).openapi("CreateProjectScore");
var patchProjectScoreSchema = objectNullish(createProjectScoreSchema).omit({ project_id: true }).openapi("PatchProjectScore");
var createProjectTagSchema = import_zod11.z.object({
  project_id: projectTagSchema.shape.project_id,
  name: projectTagSchema.shape.name,
  description: projectTagSchema.shape.description,
  color: projectTagSchema.shape.color
}).openapi("CreateProjectTag");
var patchProjectTagSchema = import_zod11.z.object({
  name: projectTagSchema.shape.name.nullish(),
  description: projectTagSchema.shape.description,
  color: projectTagSchema.shape.color
}).openapi("PatchProjectTag");
var createSpanIframeSchema = spanIframeSchema.omit({
  id: true,
  created: true,
  deleted_at: true,
  user_id: true
}).openapi("CreateSpanIFrame");
var patchSpanIframeSchema = import_zod11.z.object({
  name: spanIframeSchema.shape.name.nullish(),
  url: spanIframeSchema.shape.url.nullish(),
  post_message: spanIframeSchema.shape.post_message.nullish(),
  description: spanIframeSchema.shape.description.nullish()
}).openapi("PatchSpanIFrame");
var createViewSchema = viewSchema.omit({
  id: true,
  created: true
}).openapi("CreateView");
var patchViewSchema = import_zod11.z.object({
  object_type: viewSchema.shape.object_type,
  object_id: viewSchema.shape.object_id,
  view_type: viewSchema.shape.view_type.nullish(),
  name: viewSchema.shape.name.nullish(),
  view_data: viewSchema.shape.view_data,
  options: viewSchema.shape.options,
  user_id: viewSchema.shape.user_id
}).openapi("PatchView");
var deleteViewSchema = import_zod11.z.object({
  object_type: viewSchema.shape.object_type,
  object_id: viewSchema.shape.object_id
}).openapi("DeleteView");
var patchOrganizationSchema = import_zod11.z.object({
  name: organizationSchema.shape.name.nullish(),
  api_url: organizationSchema.shape.api_url.nullish(),
  is_universal_api: organizationSchema.shape.is_universal_api.nullish(),
  proxy_url: organizationSchema.shape.proxy_url.nullish(),
  realtime_url: organizationSchema.shape.realtime_url.nullish()
}).openapi("PatchOrganization");
var createApiKeyBaseSchema = generateBaseTableOpSchema("API key");
var createApiKeySchema = import_zod11.z.object({
  name: import_zod11.z.string().describe("Name of the api key. Does not have to be unique"),
  org_name: createApiKeyBaseSchema.shape.org_name
});
var createApiKeyOutputSchema = apiKeySchema.merge(
  import_zod11.z.object({
    key: import_zod11.z.string().describe("The raw API key. It will only be exposed this one time")
  })
).openapi("CreateApiKeyOutput");
var organizationMembersSchema = import_zod11.z.object({
  members: userSchema.pick({ id: true, email: true }).array()
}).openapi("OrganizationMembers");
var patchOrganizationMembersSchema = import_zod11.z.object({
  invite_users: import_zod11.z.object({
    ids: userSchema.shape.id.array().nullish().describe("Ids of existing users to invite"),
    emails: userSchema.shape.email.unwrap().unwrap().array().nullish().describe("Emails of users to invite"),
    send_invite_emails: import_zod11.z.boolean().nullish().describe(
      "If true, send invite emails to the users who wore actually added"
    ),
    group_ids: groupSchema.shape.id.array().nullish().describe(
      "Optional list of group ids to add newly-invited users to."
    ),
    group_names: groupSchema.shape.name.array().nullish().describe(
      "Optional list of group names to add newly-invited users to."
    ),
    group_id: groupSchema.shape.id.nullish().describe("Singular form of group_ids"),
    group_name: groupSchema.shape.name.nullish().describe("Singular form of group_names")
  }).nullish().describe("Users to invite to the organization"),
  remove_users: import_zod11.z.object({
    ids: userSchema.shape.id.array().nullish().describe("Ids of users to remove"),
    emails: userSchema.shape.email.unwrap().unwrap().array().nullish().describe("Emails of users to remove")
  }).nullish().describe("Users to remove from the organization"),
  org_name: import_zod11.z.string().nullish().describe(
    `For nearly all users, this parameter should be unnecessary. But in the rare case that your API key belongs to multiple organizations, or in case you want to explicitly assert the organization you are modifying, you may specify the name of the organization.`
  ),
  org_id: import_zod11.z.string().nullish().describe(
    `For nearly all users, this parameter should be unnecessary. But in the rare case that your API key belongs to multiple organizations, or in case you want to explicitly assert the organization you are modifying, you may specify the id of the organization.`
  )
}).openapi("PatchOrganizationMembers");
var patchOrganizationMembersOutputSchema = import_zod11.z.object({
  status: import_zod11.z.literal("success"),
  org_id: import_zod11.z.string().describe("The id of the org that was modified."),
  send_email_error: import_zod11.z.string().nullish().describe(
    "If invite emails failed to send for some reason, the patch operation will still complete, but we will return an error message here"
  )
}).openapi("PatchOrganizationMembersOutput");
var createAISecretBaseSchema = generateBaseTableOpSchema("AI Secret");
var createAISecretSchema = import_zod11.z.object({
  name: aiSecretSchema.shape.name,
  type: aiSecretSchema.shape.type,
  metadata: aiSecretSchema.shape.metadata,
  secret: import_zod11.z.string().nullish().describe(
    "Secret value. If omitted in a PUT request, the existing secret value will be left intact, not replaced with null."
  ),
  org_name: createAISecretBaseSchema.shape.org_name
}).openapi("CreateAISecret");
var deleteAISecretSchema = import_zod11.z.object({
  name: aiSecretSchema.shape.name,
  org_name: createAISecretBaseSchema.shape.org_name
}).openapi("DeleteAISecret");
var patchAISecretSchema = import_zod11.z.object({
  name: aiSecretSchema.shape.name.nullish(),
  type: aiSecretSchema.shape.type,
  metadata: aiSecretSchema.shape.metadata,
  secret: import_zod11.z.string().nullish()
}).openapi("PatchAISecret");
var createEnvVarSchema = envVarSchema.pick({ object_type: true, object_id: true, name: true }).extend({
  value: import_zod11.z.string().nullish().describe(
    "The value of the environment variable. Will be encrypted at rest."
  )
});
var patchEnvVarSchema = envVarSchema.pick({ name: true }).extend({
  value: import_zod11.z.string().nullish().describe(
    "The value of the environment variable. Will be encrypted at rest."
  )
});
var apiSpecObjectSchemas = {
  experiment: {
    object: experimentSchema,
    create: createExperimentSchema,
    patch_id: patchExperimentSchema
  },
  dataset: {
    object: datasetSchema,
    create: createDatasetSchema,
    patch_id: patchDatasetSchema
  },
  project: {
    object: projectSchema,
    create: createProjectSchema,
    patch_id: patchProjectSchema
  },
  prompt: {
    object: promptSchema,
    create: createPromptSchema,
    patch_id: patchPromptSchema
  },
  function: {
    object: functionSchema2,
    create: createFunctionSchema,
    patch_id: patchFunctionSchema
  },
  role: {
    object: roleSchema,
    create: createRoleSchema,
    patch_id: patchRoleSchema
  },
  group: {
    object: groupSchema,
    create: createGroupSchema,
    patch_id: patchGroupSchema
  },
  acl: {
    object: aclSchema,
    create: aclItemSchema,
    delete: aclItemSchema
  },
  user: {
    object: userSchema
  },
  prompt_session: {},
  project_score: {
    object: projectScoreSchema,
    create: createProjectScoreSchema,
    patch_id: patchProjectScoreSchema
  },
  project_tag: {
    object: projectTagSchema,
    create: createProjectTagSchema,
    patch_id: patchProjectTagSchema
  },
  span_iframe: {
    object: spanIframeSchema,
    create: createSpanIframeSchema,
    patch_id: patchSpanIframeSchema
  },
  view: {
    object: viewSchema,
    delete_id: deleteViewSchema,
    create: createViewSchema,
    patch_id: patchViewSchema
  },
  organization: {
    object: organizationSchema,
    patch_id: patchOrganizationSchema
  },
  api_key: {
    object: apiKeySchema,
    create: createApiKeySchema
  },
  ai_secret: {
    object: aiSecretSchema,
    create: createAISecretSchema,
    delete: deleteAISecretSchema,
    patch_id: patchAISecretSchema
  },
  env_var: {
    object: envVarSchema,
    create: createEnvVarSchema,
    patch_id: patchEnvVarSchema
  }
};

// src/string_util.ts
function capitalize(s, sep) {
  const items = sep ? s.split(sep) : [s];
  return items.map((s2) => s2 ? s2.charAt(0).toUpperCase() + s2.slice(1) : s2).join(sep || "");
}

// src/db_fields.ts
var TRANSACTION_ID_FIELD = "_xact_id";
var OBJECT_DELETE_FIELD = "_object_delete";
var IS_MERGE_FIELD = "_is_merge";
var MERGE_PATHS_FIELD = "_merge_paths";
var VALID_SOURCES = ["app", "api", "external"];

// src/span_types.ts
var spanTypeAttributeValues = [
  "llm",
  "score",
  "function",
  "eval",
  "task",
  "tool"
];

// typespecs/api_types.ts
(0, import_zod_to_openapi9.extendZodWithOpenApi)(import_zod12.z);
var auditSourcesSchema = import_zod12.z.enum(VALID_SOURCES);
var spanTypeSchema = import_zod12.z.enum(spanTypeAttributeValues).openapi("SpanType");
var spanAttributesSchema = import_zod12.z.object({
  name: import_zod12.z.string().nullish().describe("Name of the span, for display purposes only"),
  type: spanTypeSchema.nullish().describe("Type of the span, for display purposes only")
}).catchall(customTypes.unknown).describe(
  "Human-identifying attributes of the span, such as name, type, etc."
).openapi("SpanAttributes");
function generateBaseEventOpSchema(objectType) {
  const eventDescription = getEventObjectDescription(objectType);
  return import_zod12.z.object({
    id: import_zod12.z.string().describe(
      `A unique identifier for the ${eventDescription} event. If you don't provide one, BrainTrust will generate one for you`
    ),
    [TRANSACTION_ID_FIELD]: import_zod12.z.string().describe(
      `The transaction id of an event is unique to the network operation that processed the event insertion. Transaction ids are monotonically increasing over time and can be used to retrieve a versioned snapshot of the ${eventDescription} (see the \`version\` parameter)`
    ),
    created: datetimeStringSchema.describe(
      `The timestamp the ${eventDescription} event was created`
    ),
    input: customTypes.unknown,
    output: customTypes.unknown,
    expected: customTypes.unknown,
    error: customTypes.unknown.describe("The error that occurred, if any."),
    tags: import_zod12.z.array(import_zod12.z.string()).nullish().describe("A list of tags to log"),
    scores: import_zod12.z.record(import_zod12.z.number().min(0).max(1).nullish()).nullish(),
    metadata: import_zod12.z.object({
      model: import_zod12.z.string().nullish().describe("The model used for this example")
    }).catchall(customTypes.unknown).nullish().describe(
      "A dictionary with additional data about the test example, model outputs, or just about anything else that's relevant, that you can use to help find and analyze examples later. For example, you could log the `prompt`, example's `id`, or anything else that would be useful to slice/dice later. The values in `metadata` can be any JSON-serializable type, but its keys must be strings"
    ),
    metrics: import_zod12.z.object({
      start: import_zod12.z.number().nullish().describe(
        `A unix timestamp recording when the section of code which produced the ${eventDescription} event started`
      ),
      end: import_zod12.z.number().nullish().describe(
        `A unix timestamp recording when the section of code which produced the ${eventDescription} event finished`
      ),
      prompt_tokens: import_zod12.z.number().int().nullish().describe(
        `The number of tokens in the prompt used to generate the ${eventDescription} event (only set if this is an LLM span)`
      ),
      completion_tokens: import_zod12.z.number().int().nullish().describe(
        `The number of tokens in the completion generated by the model (only set if this is an LLM span)`
      ),
      tokens: import_zod12.z.number().int().nullish().describe(
        `The total number of tokens in the input and output of the ${eventDescription} event.`
      ),
      // Legacy non-numerical metrics.
      caller_functionname: customTypes.unknown.describe(
        "This metric is deprecated"
      ),
      caller_filename: customTypes.unknown.describe(
        "This metric is deprecated"
      ),
      caller_lineno: customTypes.unknown.describe(
        "This metric is deprecated"
      )
    }).catchall(import_zod12.z.number()).nullish().describe(
      `Metrics are numerical measurements tracking the execution of the code that produced the ${eventDescription} event. Use "start" and "end" to track the time span over which the ${eventDescription} event was produced`
    ),
    context: import_zod12.z.object({
      caller_functionname: import_zod12.z.string().nullish().describe(
        `The function in code which created the ${eventDescription} event`
      ),
      caller_filename: import_zod12.z.string().nullish().describe(
        `Name of the file in code where the ${eventDescription} event was created`
      ),
      caller_lineno: import_zod12.z.number().int().nullish().describe(
        `Line of code where the ${eventDescription} event was created`
      )
    }).catchall(customTypes.unknown).nullish().describe(
      `Context is additional information about the code that produced the ${eventDescription} event. It is essentially the textual counterpart to \`metrics\`. Use the \`caller_*\` attributes to track the location in code which produced the ${eventDescription} event`
    ),
    span_id: import_zod12.z.string().describe(
      `A unique identifier used to link different ${eventDescription} events together as part of a full trace. See the [tracing guide](https://www.braintrust.dev/docs/guides/tracing) for full details on tracing`
    ),
    span_parents: import_zod12.z.string().array().nullish().describe(
      `An array of the parent \`span_ids\` of this ${eventDescription} event. This should be empty for the root span of a trace, and should most often contain just one parent element for subspans`
    ),
    root_span_id: import_zod12.z.string().describe(
      `A unique identifier for the trace this ${eventDescription} event belongs to`
    ),
    span_attributes: spanAttributesSchema.nullish(),
    origin: objectReferenceSchema.nullish().describe("Indicates the event was copied from another object."),
    [OBJECT_DELETE_FIELD]: import_zod12.z.boolean().nullish().describe(
      `Pass \`${OBJECT_DELETE_FIELD}=true\` to mark the ${eventDescription} event deleted. Deleted events will not show up in subsequent fetches for this ${eventDescription}`
    ),
    is_root: import_zod12.z.boolean().nullish().describe(`Whether this span is a root span`)
  });
}
function generateBaseEventFeedbackSchema(objectType) {
  const eventObjectType2 = getEventObjectType(objectType);
  const eventDescription = getEventObjectDescription(objectType);
  return import_zod12.z.object({
    id: import_zod12.z.string().describe(
      `The id of the ${eventDescription} event to log feedback for. This is the row \`id\` returned by \`POST /v1/${eventObjectType2}/{${objectType}_id}/insert\``
    ),
    scores: import_zod12.z.record(import_zod12.z.number().min(0).max(1).nullish()).nullish().describe(
      `A dictionary of numeric values (between 0 and 1) to log. These scores will be merged into the existing scores for the ${eventDescription} event`
    ),
    expected: customTypes.unknown.describe(
      "The ground truth value (an arbitrary, JSON serializable object) that you'd compare to `output` to determine if your `output` value is correct or not"
    ),
    tags: import_zod12.z.array(import_zod12.z.string()).nullish().describe("A list of tags to log"),
    comment: import_zod12.z.string().nullish().describe(
      `An optional comment string to log about the ${eventDescription} event`
    ),
    metadata: import_zod12.z.record(customTypes.unknown).nullish().describe(
      "A dictionary with additional data about the feedback. If you have a `user_id`, you can log it here and access it in the Braintrust UI. Note, this metadata does not correspond to the main event itself, but rather the audit log attached to the event."
    ),
    source: auditSourcesSchema.nullish().describe(
      'The source of the feedback. Must be one of "external" (default), "app", or "api"'
    )
  });
}
var fetchLimitParamSchema = import_zod12.z.coerce.number().int().nonnegative().describe(
  [
    "limit the number of traces fetched",
    `Fetch queries may be paginated if the total result size is expected to be large (e.g. project_logs which accumulate over a long time). Note that fetch queries only support pagination in descending time order (from latest to earliest \`${TRANSACTION_ID_FIELD}\`. Furthermore, later pages may return rows which showed up in earlier pages, except with an earlier \`${TRANSACTION_ID_FIELD}\`. This happens because pagination occurs over the whole version history of the event log. You will most likely want to exclude any such duplicate, outdated rows (by \`id\`) from your combined result set.`,
    `The \`limit\` parameter controls the number of full traces to return. So you may end up with more individual rows than the specified limit if you are fetching events containing traces.`
  ].join("\n\n")
).openapi("FetchLimit");
var fetchPaginationCursorDescription = [
  "DEPRECATION NOTICE: The manually-constructed pagination cursor is deprecated in favor of the explicit 'cursor' returned by object fetch requests. Please prefer the 'cursor' argument going forwards.",
  "Together, `max_xact_id` and `max_root_span_id` form a pagination cursor",
  `Since a paginated fetch query returns results in order from latest to earliest, the cursor for the next page can be found as the row with the minimum (earliest) value of the tuple \`(${TRANSACTION_ID_FIELD}, root_span_id)\`. See the documentation of \`limit\` for an overview of paginating fetch queries.`
].join("\n\n");
var maxXactIdSchema = import_zod12.z.string().describe(fetchPaginationCursorDescription).openapi("MaxXactId");
var maxRootSpanIdSchema = import_zod12.z.string().describe(fetchPaginationCursorDescription).openapi("MaxRootSpanId");
var fetchPaginationCursorSchema = import_zod12.z.string().describe(
  [
    "An opaque string to be used as a cursor for the next page of results, in order from latest to earliest.",
    "The string can be obtained directly from the `cursor` property of the previous fetch query"
  ].join("\n\n")
).openapi("FetchPaginationCursor");
var versionSchema = import_zod12.z.string().describe(
  [
    "Retrieve a snapshot of events from a past time",
    "The version id is essentially a filter on the latest event transaction id. You can use the `max_xact_id` returned by a past fetch as the version to reproduce that exact fetch."
  ].join("\n\n")
).openapi("Version");
var fetchEventsRequestSchema = import_zod12.z.object({
  limit: fetchLimitParamSchema.nullish(),
  cursor: fetchPaginationCursorSchema.nullish(),
  max_xact_id: maxXactIdSchema.nullish(),
  max_root_span_id: maxRootSpanIdSchema.nullish(),
  version: versionSchema.nullish()
}).openapi("FetchEventsRequest");
function makeFetchEventsResponseSchema(objectType, eventSchema) {
  const eventName = capitalize(getEventObjectType(objectType), "_").replace(
    "_",
    ""
  );
  return import_zod12.z.object({
    events: eventSchema.array().describe("A list of fetched events"),
    cursor: import_zod12.z.string().nullish().describe(
      [
        "Pagination cursor",
        "Pass this string directly as the `cursor` param to your next fetch request to get the next page of results. Not provided if the returned result set is empty."
      ].join("\n\n")
    )
  }).openapi(`Fetch${eventName}EventsResponse`);
}
var experimentEventBaseSchema = generateBaseEventOpSchema("experiment");
var experimentEventSchema = import_zod12.z.object({
  id: experimentEventBaseSchema.shape.id,
  [TRANSACTION_ID_FIELD]: experimentEventBaseSchema.shape[TRANSACTION_ID_FIELD],
  created: experimentEventBaseSchema.shape.created,
  project_id: experimentSchema.shape.project_id,
  experiment_id: experimentSchema.shape.id,
  input: experimentEventBaseSchema.shape.input.describe(
    "The arguments that uniquely define a test case (an arbitrary, JSON serializable object). Later on, Braintrust will use the `input` to know whether two test cases are the same between experiments, so they should not contain experiment-specific state. A simple rule of thumb is that if you run the same experiment twice, the `input` should be identical"
  ),
  output: experimentEventBaseSchema.shape.output.describe(
    "The output of your application, including post-processing (an arbitrary, JSON serializable object), that allows you to determine whether the result is correct or not. For example, in an app that generates SQL queries, the `output` should be the _result_ of the SQL query generated by the model, not the query itself, because there may be multiple valid queries that answer a single question"
  ),
  expected: experimentEventBaseSchema.shape.expected.describe(
    "The ground truth value (an arbitrary, JSON serializable object) that you'd compare to `output` to determine if your `output` value is correct or not. Braintrust currently does not compare `output` to `expected` for you, since there are so many different ways to do that correctly. Instead, these values are just used to help you navigate your experiments while digging into analyses. However, we may later use these values to re-score outputs or fine-tune your models"
  ),
  error: experimentEventBaseSchema.shape.error,
  scores: experimentEventBaseSchema.shape.scores.describe(
    "A dictionary of numeric values (between 0 and 1) to log. The scores should give you a variety of signals that help you determine how accurate the outputs are compared to what you expect and diagnose failures. For example, a summarization app might have one score that tells you how accurate the summary is, and another that measures the word similarity between the generated and grouth truth summary. The word similarity score could help you determine whether the summarization was covering similar concepts or not. You can use these scores to help you sort, filter, and compare experiments"
  ),
  metadata: experimentEventBaseSchema.shape.metadata,
  tags: experimentEventBaseSchema.shape.tags,
  metrics: experimentEventBaseSchema.shape.metrics,
  context: experimentEventBaseSchema.shape.context,
  span_id: experimentEventBaseSchema.shape.span_id,
  span_parents: experimentEventBaseSchema.shape.span_parents,
  root_span_id: experimentEventBaseSchema.shape.root_span_id,
  span_attributes: experimentEventBaseSchema.shape.span_attributes,
  is_root: experimentEventBaseSchema.shape.is_root,
  origin: experimentEventBaseSchema.shape.origin
}).openapi("ExperimentEvent");
var datasetEventBaseSchema = generateBaseEventOpSchema("dataset");
var datasetEventSchema = import_zod12.z.object({
  id: datasetEventBaseSchema.shape.id,
  [TRANSACTION_ID_FIELD]: datasetEventBaseSchema.shape[TRANSACTION_ID_FIELD],
  created: datasetEventBaseSchema.shape.created,
  project_id: datasetSchema.shape.project_id,
  dataset_id: datasetSchema.shape.id,
  input: datasetEventBaseSchema.shape.input.describe(
    "The argument that uniquely define an input case (an arbitrary, JSON serializable object)"
  ),
  expected: datasetEventBaseSchema.shape.expected.describe(
    "The output of your application, including post-processing (an arbitrary, JSON serializable object)"
  ),
  metadata: datasetEventBaseSchema.shape.metadata,
  tags: datasetEventBaseSchema.shape.tags,
  span_id: datasetEventBaseSchema.shape.span_id,
  root_span_id: datasetEventBaseSchema.shape.root_span_id,
  is_root: datasetEventBaseSchema.shape.is_root,
  origin: datasetEventBaseSchema.shape.origin
}).openapi("DatasetEvent");
var promptSessionEventBaseSchema = generateBaseEventOpSchema("prompt_session");
var promptSessionEventSchema = import_zod12.z.object({
  id: promptSessionEventBaseSchema.shape.id,
  [TRANSACTION_ID_FIELD]: promptSessionEventBaseSchema.shape[TRANSACTION_ID_FIELD],
  created: promptSessionEventBaseSchema.shape.created,
  project_id: promptSchema.shape.project_id,
  prompt_session_id: promptSchema.shape.id,
  prompt_session_data: customTypes.unknown.describe(
    "Data about the prompt session"
  ),
  prompt_data: customTypes.unknown.describe("Data about the prompt"),
  function_data: customTypes.unknown.describe("Data about the function"),
  object_data: customTypes.unknown.describe("Data about the mapped data"),
  completion: customTypes.unknown.describe("Data about the completion"),
  tags: promptSessionEventBaseSchema.shape.tags
}).openapi("PromptSessionEvent");
var projectLogsEventBaseSchema = generateBaseEventOpSchema("project");
var projectLogsLogIdLiteralSchema = import_zod12.z.literal("g").describe("A literal 'g' which identifies the log as a project log");
var projectLogsEventSchema = import_zod12.z.object({
  id: projectLogsEventBaseSchema.shape.id,
  [TRANSACTION_ID_FIELD]: projectLogsEventBaseSchema.shape[TRANSACTION_ID_FIELD],
  created: projectLogsEventBaseSchema.shape.created,
  org_id: projectSchema.shape.org_id,
  project_id: projectSchema.shape.id,
  log_id: projectLogsLogIdLiteralSchema,
  input: projectLogsEventBaseSchema.shape.input.describe(
    "The arguments that uniquely define a user input (an arbitrary, JSON serializable object)."
  ),
  output: projectLogsEventBaseSchema.shape.output.describe(
    "The output of your application, including post-processing (an arbitrary, JSON serializable object), that allows you to determine whether the result is correct or not. For example, in an app that generates SQL queries, the `output` should be the _result_ of the SQL query generated by the model, not the query itself, because there may be multiple valid queries that answer a single question."
  ),
  expected: projectLogsEventBaseSchema.shape.expected.describe(
    "The ground truth value (an arbitrary, JSON serializable object) that you'd compare to `output` to determine if your `output` value is correct or not. Braintrust currently does not compare `output` to `expected` for you, since there are so many different ways to do that correctly. Instead, these values are just used to help you navigate while digging into analyses. However, we may later use these values to re-score outputs or fine-tune your models."
  ),
  error: projectLogsEventBaseSchema.shape.error,
  scores: projectLogsEventBaseSchema.shape.scores.describe(
    "A dictionary of numeric values (between 0 and 1) to log. The scores should give you a variety of signals that help you determine how accurate the outputs are compared to what you expect and diagnose failures. For example, a summarization app might have one score that tells you how accurate the summary is, and another that measures the word similarity between the generated and grouth truth summary. The word similarity score could help you determine whether the summarization was covering similar concepts or not. You can use these scores to help you sort, filter, and compare logs."
  ),
  metadata: projectLogsEventBaseSchema.shape.metadata,
  tags: projectLogsEventBaseSchema.shape.tags,
  metrics: projectLogsEventBaseSchema.shape.metrics,
  context: projectLogsEventBaseSchema.shape.context,
  span_id: projectLogsEventBaseSchema.shape.span_id,
  span_parents: projectLogsEventBaseSchema.shape.span_parents,
  root_span_id: projectLogsEventBaseSchema.shape.root_span_id,
  is_root: projectLogsEventBaseSchema.shape.is_root,
  span_attributes: projectLogsEventBaseSchema.shape.span_attributes,
  origin: projectLogsEventBaseSchema.shape.origin
}).openapi("ProjectLogsEvent");
var spanIdsDescription = [
  "Use `span_id`, `root_span_id`, and `span_parents` instead of `_parent_id`, which is now deprecated. The span_id is a unique identifier describing the row's place in the a trace, and the root_span_id is a unique identifier for the whole trace. See the [guide](https://www.braintrust.dev/docs/guides/tracing) for full details.",
  'For example, say we have logged a row `{"id": "abc", "span_id": "span0", "root_span_id": "root_span0", "input": "foo", "output": "bar", "expected": "boo", "scores": {"correctness": 0.33}}`. We can create a sub-span of the parent row by logging `{"id": "llm_call", "span_id": "span1", "root_span_id": "root_span0", "span_parents": ["span0"], "input": {"prompt": "What comes after foo?"}, "output": "bar", "metrics": {"tokens": 1}}`. In the webapp, only the root span row `"abc"` will show up in the summary view. You can view the full trace hierarchy (in this case, the `"llm_call"` row) by clicking on the "abc" row.',
  "If the row is being merged into an existing row, this field will be ignored."
].join("\n\n");
var insertSystemControlFieldsSchema = import_zod12.z.object({
  [IS_MERGE_FIELD]: import_zod12.z.boolean().nullish().describe(
    [
      "The `_is_merge` field controls how the row is merged with any existing row with the same id in the DB. By default (or when set to `false`), the existing row is completely replaced by the new row. When set to `true`, the new row is deep-merged into the existing row, if one is found. If no existing row is found, the new row is inserted as is.",
      'For example, say there is an existing row in the DB `{"id": "foo", "input": {"a": 5, "b": 10}}`. If we merge a new row as `{"_is_merge": true, "id": "foo", "input": {"b": 11, "c": 20}}`, the new row will be `{"id": "foo", "input": {"a": 5, "b": 11, "c": 20}}`. If we replace the new row as `{"id": "foo", "input": {"b": 11, "c": 20}}`, the new row will be `{"id": "foo", "input": {"b": 11, "c": 20}}`'
    ].join("\n\n")
  ),
  [MERGE_PATHS_FIELD]: import_zod12.z.string().array().array().nullish().describe(
    [
      "The `_merge_paths` field allows controlling the depth of the merge, when `_is_merge=true`. `_merge_paths` is a list of paths, where each path is a list of field names. The deep merge will not descend below any of the specified merge paths.",
      'For example, say there is an existing row in the DB `{"id": "foo", "input": {"a": {"b": 10}, "c": {"d": 20}}, "output": {"a": 20}}`. If we merge a new row as `{"_is_merge": true, "_merge_paths": [["input", "a"], ["output"]], "input": {"a": {"q": 30}, "c": {"e": 30}, "bar": "baz"}, "output": {"d": 40}}`, the new row will be `{"id": "foo": "input": {"a": {"q": 30}, "c": {"d": 20, "e": 30}, "bar": "baz"}, "output": {"d": 40}}`. In this case, due to the merge paths, we have replaced `input.a` and `output`, but have still deep-merged `input` and `input.c`.'
    ].join("\n\n")
  ),
  _parent_id: import_zod12.z.string().nullish().describe(
    [
      "DEPRECATED: The `_parent_id` field is deprecated and should not be used. Support for `_parent_id` will be dropped in a future version of Braintrust. Log `span_id`, `root_span_id`, and `span_parents` explicitly instead.",
      "Use the `_parent_id` field to create this row as a subspan of an existing row. Tracking hierarchical relationships are important for tracing (see the [guide](https://www.braintrust.dev/docs/guides/tracing) for full details).",
      'For example, say we have logged a row `{"id": "abc", "input": "foo", "output": "bar", "expected": "boo", "scores": {"correctness": 0.33}}`. We can create a sub-span of the parent row by logging `{"_parent_id": "abc", "id": "llm_call", "input": {"prompt": "What comes after foo?"}, "output": "bar", "metrics": {"tokens": 1}}`. In the webapp, only the root span row `"abc"` will show up in the summary view. You can view the full trace hierarchy (in this case, the `"llm_call"` row) by clicking on the "abc" row.',
      "If the row is being merged into an existing row, this field will be ignored."
    ].join("\n\n")
  ),
  span_id: import_zod12.z.string().nullish().describe(spanIdsDescription),
  root_span_id: import_zod12.z.string().nullish().describe(spanIdsDescription),
  span_parents: import_zod12.z.string().array().nullish().describe(spanIdsDescription)
});
function makeInsertEventSchemas(objectType, insertSchema) {
  const eventDescription = getEventObjectDescription(objectType);
  const article = getObjectArticle(objectType);
  const eventSchemaName = capitalize(
    getEventObjectType(objectType),
    "_"
  ).replace("_", "");
  const eventSchema = insertSchema.merge(insertSystemControlFieldsSchema).describe(`${capitalize(article)} ${eventDescription} event`).openapi(`Insert${eventSchemaName}Event`);
  const requestSchema = import_zod12.z.object({
    events: eventSchema.array().describe(`A list of ${eventDescription} events to insert`)
  }).openapi(`Insert${eventSchemaName}EventRequest`);
  return { eventSchema, requestSchema };
}
var insertEventsResponseSchema = import_zod12.z.object({
  row_ids: import_zod12.z.string().array().describe(
    "The ids of all rows that were inserted, aligning one-to-one with the rows provided as input"
  )
}).openapi("InsertEventsResponse");
var feedbackResponseSchema = import_zod12.z.object({
  status: import_zod12.z.literal("success")
}).openapi("FeedbackResponseSchema");
var insertExperimentEventBaseSchema = objectNullish(
  experimentEventSchema.pick({
    input: true,
    output: true,
    expected: true,
    error: true,
    scores: true,
    metadata: true,
    tags: true,
    metrics: true,
    context: true,
    span_attributes: true,
    id: true,
    created: true,
    origin: true
  }).extend({
    [OBJECT_DELETE_FIELD]: experimentEventBaseSchema.shape[OBJECT_DELETE_FIELD]
  })
);
var {
  eventSchema: insertExperimentEventSchema,
  requestSchema: insertExperimentEventsRequestSchema
} = makeInsertEventSchemas("experiment", insertExperimentEventBaseSchema);
var insertDatasetEventBaseSchema = objectNullish(
  datasetEventSchema.pick({
    input: true,
    expected: true,
    metadata: true,
    tags: true,
    id: true,
    created: true,
    origin: true
  }).extend({
    [OBJECT_DELETE_FIELD]: datasetEventBaseSchema.shape[OBJECT_DELETE_FIELD]
  })
);
var {
  eventSchema: insertDatasetEventSchema,
  requestSchema: insertDatasetEventsRequestSchema
} = makeInsertEventSchemas("dataset", insertDatasetEventBaseSchema);
var insertProjectLogsEventBaseSchema = objectNullish(
  projectLogsEventSchema.pick({
    input: true,
    output: true,
    expected: true,
    error: true,
    scores: true,
    metadata: true,
    tags: true,
    metrics: true,
    context: true,
    span_attributes: true,
    id: true,
    created: true,
    origin: true
  }).extend({
    [OBJECT_DELETE_FIELD]: projectLogsEventBaseSchema.shape[OBJECT_DELETE_FIELD]
  })
);
var {
  eventSchema: insertProjectLogsEventSchema,
  requestSchema: insertProjectLogsEventsRequestSchema
} = makeInsertEventSchemas("project", insertProjectLogsEventBaseSchema);
function makeFeedbackRequestSchema(objectType, feedbackSchema) {
  const eventDescription = getEventObjectDescription(objectType);
  const eventSchemaName = capitalize(
    getEventObjectType(objectType),
    "_"
  ).replace("_", "");
  return import_zod12.z.object({
    feedback: feedbackSchema.array().describe(`A list of ${eventDescription} feedback items`)
  }).openapi(`Feedback${eventSchemaName}EventRequest`);
}
var feedbackExperimentRequestBaseSchema = generateBaseEventFeedbackSchema("experiment");
var feedbackExperimentItemSchema = feedbackExperimentRequestBaseSchema.pick({
  id: true,
  scores: true,
  expected: true,
  comment: true,
  metadata: true,
  source: true,
  tags: true
}).openapi("FeedbackExperimentItem");
var feedbackExperimentRequestSchema = makeFeedbackRequestSchema(
  "experiment",
  feedbackExperimentItemSchema
);
var feedbackDatasetRequestBaseSchema = generateBaseEventFeedbackSchema("dataset");
var feedbackDatasetItemSchema = feedbackDatasetRequestBaseSchema.pick({
  id: true,
  comment: true,
  metadata: true,
  source: true,
  tags: true
}).openapi("FeedbackDatasetItem");
var feedbackDatasetRequestSchema = makeFeedbackRequestSchema(
  "dataset",
  feedbackDatasetItemSchema
);
var feedbackProjectLogsRequestBaseSchema = generateBaseEventFeedbackSchema("project");
var feedbackProjectLogsItemSchema = feedbackProjectLogsRequestBaseSchema.pick({
  id: true,
  scores: true,
  expected: true,
  comment: true,
  metadata: true,
  source: true,
  tags: true
}).openapi("FeedbackProjectLogsItem");
var feedbackProjectLogsRequestSchema = makeFeedbackRequestSchema(
  "project",
  feedbackProjectLogsItemSchema
);
var apiSpecEventObjectSchemas = {
  experiment: {
    event: experimentEventSchema,
    fetchResponse: makeFetchEventsResponseSchema(
      "experiment",
      experimentEventSchema
    ),
    insertEvent: insertExperimentEventSchema,
    insertRequest: insertExperimentEventsRequestSchema,
    feedbackItem: feedbackExperimentItemSchema,
    feedbackRequest: feedbackExperimentRequestSchema
  },
  dataset: {
    event: datasetEventSchema,
    fetchResponse: makeFetchEventsResponseSchema("dataset", datasetEventSchema),
    insertEvent: insertDatasetEventSchema,
    insertRequest: insertDatasetEventsRequestSchema,
    feedbackItem: feedbackDatasetItemSchema,
    feedbackRequest: feedbackDatasetRequestSchema
  },
  project_logs: {
    event: projectLogsEventSchema,
    fetchResponse: makeFetchEventsResponseSchema(
      "project",
      projectLogsEventSchema
    ),
    insertEvent: insertProjectLogsEventSchema,
    insertRequest: insertProjectLogsEventsRequestSchema,
    feedbackItem: feedbackProjectLogsItemSchema,
    feedbackRequest: feedbackProjectLogsRequestSchema
  },
  prompt: {
    event: promptSchema
  },
  function: {
    event: functionSchema2
  },
  prompt_session: {}
};
function makeCrossObjectIndividualRequestSchema(objectType) {
  const eventObjectType2 = getEventObjectType(objectType);
  const eventDescription = getEventObjectDescription(objectType);
  const eventObjectSchema = apiSpecEventObjectSchemas[eventObjectType2];
  const insertObject = import_zod12.z.object({
    ...eventObjectSchema.insertEvent ? {
      events: eventObjectSchema.insertEvent.array().nullish().describe(`A list of ${eventDescription} events to insert`)
    } : {},
    ...eventObjectSchema.feedbackItem ? {
      feedback: eventObjectSchema.feedbackItem.array().nullish().describe(`A list of ${eventDescription} feedback items`)
    } : {}
  });
  return import_zod12.z.record(import_zod12.z.string().uuid(), insertObject).nullish().describe(
    `A mapping from ${objectType} id to a set of log events and feedback items to insert`
  );
}
function makeCrossObjectIndividualResponseSchema(objectType) {
  return import_zod12.z.record(import_zod12.z.string().uuid(), insertEventsResponseSchema).nullish().describe(
    `A mapping from ${objectType} id to row ids for inserted \`events\``
  );
}
var crossObjectInsertRequestSchema = import_zod12.z.object({
  experiment: makeCrossObjectIndividualRequestSchema("experiment"),
  dataset: makeCrossObjectIndividualRequestSchema("dataset"),
  project_logs: makeCrossObjectIndividualRequestSchema("project")
}).openapi("CrossObjectInsertRequest");
var crossObjectInsertResponseSchema = import_zod12.z.object({
  experiment: makeCrossObjectIndividualResponseSchema("experiment"),
  dataset: makeCrossObjectIndividualResponseSchema("dataset"),
  project_logs: makeCrossObjectIndividualResponseSchema("project")
}).openapi("CrossObjectInsertResponse");
var summarizeScoresParamSchema = import_zod12.z.coerce.boolean().describe(
  "Whether to summarize the scores and metrics. If false (or omitted), only the metadata will be returned."
);
var comparisonExperimentIdParamSchema = import_zod12.z.string().uuid().describe(
  "The experiment to compare against, if summarizing scores and metrics. If omitted, will fall back to the `base_exp_id` stored in the experiment metadata, and then to the most recent experiment run in the same project. Must pass `summarize_scores=true` for this id to be used"
);
var summarizeDataParamSchema = import_zod12.z.coerce.boolean().describe(
  "Whether to summarize the data. If false (or omitted), only the metadata will be returned."
);
var summarizeExperimentResponseSchema = import_zod12.z.object({
  project_name: import_zod12.z.string().describe("Name of the project that the experiment belongs to"),
  experiment_name: import_zod12.z.string().describe("Name of the experiment"),
  project_url: import_zod12.z.string().url().describe("URL to the project's page in the Braintrust app"),
  experiment_url: import_zod12.z.string().url().describe("URL to the experiment's page in the Braintrust app"),
  comparison_experiment_name: import_zod12.z.string().nullish().describe("The experiment which scores are baselined against"),
  scores: import_zod12.z.record(
    import_zod12.z.object({
      name: import_zod12.z.string().describe("Name of the score"),
      score: import_zod12.z.number().min(0).max(1).describe("Average score across all examples"),
      diff: import_zod12.z.number().min(-1).max(1).optional().describe(
        "Difference in score between the current and comparison experiment"
      ),
      improvements: import_zod12.z.number().int().min(0).describe("Number of improvements in the score"),
      regressions: import_zod12.z.number().int().min(0).describe("Number of regressions in the score")
    }).describe("Summary of a score's performance").openapi("ScoreSummary")
  ).nullish().describe("Summary of the experiment's scores"),
  metrics: import_zod12.z.record(
    import_zod12.z.object({
      name: import_zod12.z.string().describe("Name of the metric"),
      metric: import_zod12.z.number().describe("Average metric across all examples"),
      unit: import_zod12.z.string().describe("Unit label for the metric"),
      diff: import_zod12.z.number().optional().describe(
        "Difference in metric between the current and comparison experiment"
      ),
      improvements: import_zod12.z.number().int().min(0).describe("Number of improvements in the metric"),
      regressions: import_zod12.z.number().int().min(0).describe("Number of regressions in the metric")
    }).describe("Summary of a metric's performance").openapi("MetricSummary")
  ).nullish().describe("Summary of the experiment's metrics")
}).describe("Summary of an experiment").openapi("SummarizeExperimentResponse");
var summarizeDatasetResponseSchema = import_zod12.z.object({
  project_name: import_zod12.z.string().describe("Name of the project that the dataset belongs to"),
  dataset_name: import_zod12.z.string().describe("Name of the dataset"),
  project_url: import_zod12.z.string().url().describe("URL to the project's page in the Braintrust app"),
  dataset_url: import_zod12.z.string().url().describe("URL to the dataset's page in the Braintrust app"),
  data_summary: import_zod12.z.object({
    total_records: import_zod12.z.number().int().min(0).describe("Total number of records in the dataset")
  }).nullish().describe("Summary of a dataset's data").openapi("DataSummary")
}).describe("Summary of a dataset").openapi("SummarizeDatasetResponse");
var objectTypeSummarizeResponseSchemas = {
  experiment: summarizeExperimentResponseSchema,
  dataset: summarizeDatasetResponseSchema
};
var asyncScoringStateSchema = import_zod12.z.union([
  import_zod12.z.object({
    status: import_zod12.z.literal("enabled"),
    token: import_zod12.z.string(),
    function_ids: import_zod12.z.array(functionIdSchema).nonempty()
  }),
  // Explicitly disabled.
  import_zod12.z.object({
    status: import_zod12.z.literal("disabled")
  }),
  // Inactive but may be selected later.
  import_zod12.z.null()
]).openapi("AsyncScoringState");
var asyncScoringControlSchema = import_zod12.z.discriminatedUnion("kind", [
  import_zod12.z.object({
    kind: import_zod12.z.literal("score_update"),
    token: import_zod12.z.string()
  }),
  import_zod12.z.object({
    kind: import_zod12.z.literal("state_override"),
    state: asyncScoringStateSchema
  }),
  import_zod12.z.object({
    kind: import_zod12.z.literal("state_force_reselect")
  })
]).openapi("AsyncScoringControl");

// typespecs/attachment-reference.ts
var import_zod13 = require("zod");
var BRAINTRUST_ATTACHMENT = "braintrust_attachment";
var EXTERNAL_ATTACHMENT = "external_attachment";
var braintrustAttachmentReferenceSchema = import_zod13.z.object({
  type: import_zod13.z.literal(BRAINTRUST_ATTACHMENT).describe("An identifier to help disambiguate parsing."),
  filename: import_zod13.z.string().min(1).describe(
    "Human-readable filename for user interfaces. Not related to attachment storage."
  ),
  content_type: import_zod13.z.string().min(1).describe("MIME type of this file."),
  key: import_zod13.z.string().min(1).describe("Key in the object store bucket for this attachment.")
}).openapi("BraintrustAttachmentReference");
var externalAttachmentReferenceSchema = import_zod13.z.object({
  type: import_zod13.z.literal(EXTERNAL_ATTACHMENT).describe("An identifier to help disambiguate parsing."),
  filename: import_zod13.z.string().min(1).describe(
    "Human-readable filename for user interfaces. Not related to attachment storage."
  ),
  content_type: import_zod13.z.string().min(1).describe("MIME type of this file."),
  url: import_zod13.z.string().min(1).describe(
    "Fully qualified URL to the object in the external object store."
  )
}).openapi("ExternalAttachmentReference");
var attachmentReferenceSchema = import_zod13.z.discriminatedUnion("type", [
  braintrustAttachmentReferenceSchema,
  externalAttachmentReferenceSchema
]).openapi("AttachmentReference");
var uploadStatusSchema = import_zod13.z.enum(["uploading", "done", "error"]).openapi("UploadStatus");
var attachmentStatusSchema = import_zod13.z.object({
  upload_status: uploadStatusSchema,
  error_message: import_zod13.z.string().nullish().transform((x) => x || void 0).describe("Describes the error encountered while uploading.")
}).openapi("AttachmentStatus");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BRAINTRUST_ATTACHMENT,
  BRAINTRUST_PARAMS,
  DEFAULT_IF_EXISTS,
  EXTERNAL_ATTACHMENT,
  aclBatchUpdateRequestSchema,
  aclBatchUpdateResponseSchema,
  aclItemSchema,
  aclObjectTypeEnum,
  aclSchema,
  aiSecretSchema,
  apiKeySchema,
  apiSpecEventObjectSchemas,
  apiSpecObjectSchemas,
  appLimitParamSchema,
  asyncScoringControlSchema,
  asyncScoringStateSchema,
  attachmentReferenceSchema,
  attachmentStatusSchema,
  baseSSEEventSchema,
  braintrustAttachmentReferenceSchema,
  callEventSchema,
  chatCompletionContentPartImageSchema,
  chatCompletionContentPartSchema,
  chatCompletionContentPartTextSchema,
  chatCompletionMessageParamSchema,
  chatCompletionMessageToolCallSchema,
  codeBundleSchema,
  comparisonExperimentIdParamSchema,
  createAISecretSchema,
  createApiKeyOutputSchema,
  createApiKeySchema,
  createDatasetSchema,
  createEnvVarSchema,
  createExperimentSchema,
  createFunctionSchema,
  createGroupSchema,
  createProjectSchema,
  createProjectScoreSchema,
  createProjectTagSchema,
  createPromptSchema,
  createSpanIframeSchema,
  createViewSchema,
  crossObjectInsertRequestSchema,
  crossObjectInsertResponseSchema,
  customColumnSchema,
  customTypes,
  customTypesMode,
  datasetEventSchema,
  datasetSchema,
  datetimeStringSchema,
  deleteAISecretSchema,
  endingBeforeSchema,
  envVarObjectTypeEnum,
  envVarSchema,
  eventObjectType,
  experimentEventSchema,
  experimentSchema,
  extendedSavedFunctionIdSchema,
  externalAttachmentReferenceSchema,
  feedbackResponseSchema,
  fetchEventsRequestSchema,
  fetchLimitParamSchema,
  fetchPaginationCursorSchema,
  functionDataSchema,
  functionFormatEnum,
  functionIdSchema,
  functionObjectTypeEnum,
  functionOutputTypeEnum,
  functionSchema,
  functionTypeEnum,
  getEventObjectDescription,
  getEventObjectType,
  getObjectArticle,
  gitFieldsSchema,
  gitMetadataSettingsSchema,
  groupSchema,
  ifExistsEnum,
  insertEventsResponseSchema,
  invokeApiSchema,
  invokeFunctionNonIdArgsSchema,
  invokeFunctionSchema,
  invokeParent,
  jsonSchema,
  literalSchema,
  makeObjectIdsFilterSchema,
  maxOverWindowSchema,
  maxRootSpanIdSchema,
  maxXactIdSchema,
  memberSchema,
  messageRoleSchema,
  modelParamsSchema,
  objectReferenceSchema,
  objectTypeSummarizeResponseSchemas,
  objectTypes,
  objectTypesWithEvent,
  onlineScoreConfigSchema,
  organizationMembersSchema,
  organizationSchema,
  patchAISecretSchema,
  patchDatasetSchema,
  patchEnvVarSchema,
  patchExperimentSchema,
  patchGroupSchema,
  patchOrganizationMembersOutputSchema,
  patchOrganizationMembersSchema,
  patchOrganizationSchema,
  patchProjectSchema,
  patchProjectScoreSchema,
  patchProjectTagSchema,
  patchPromptSchema,
  patchRoleSchema,
  patchSpanIframeSchema,
  patchViewSchema,
  permissionEnum,
  playgroundLogsLogIdLiteralSchema,
  projectLogsEventSchema,
  projectLogsLogIdLiteralSchema,
  projectSchema,
  projectScoreCategory,
  projectScoreSchema,
  projectScoreTypeEnum,
  projectSettingsSchema,
  projectTagSchema,
  promptBaseSchema,
  promptBlockDataSchema,
  promptDataSchema,
  promptLogIdLiteralSchema,
  promptOptionsSchema,
  promptSchema,
  promptSessionEventSchema,
  promptSessionSchema,
  repoInfoSchema,
  resourcesSchema,
  responseFormatJsonSchemaSchema,
  responseFormatSchema,
  roleSchema,
  runEvalSchema,
  runtimeContextSchema,
  savedFunctionIdSchema,
  scoreSchema,
  spanAttributesSchema,
  spanFieldOrderItem,
  spanIframeBaseSchema,
  spanIframeSchema,
  spanTypeSchema,
  sseConsoleEventDataSchema,
  sseConsoleEventSchema,
  sseDataEventSchema,
  sseDoneEventSchema,
  sseErrorEventSchema,
  sseProgressEventDataSchema,
  sseProgressEventSchema,
  sseStartEventSchema,
  sseTextEventSchema,
  startingAfterSchema,
  streamingModeEnum,
  strictPromptDataSchema,
  summarizeDataParamSchema,
  summarizeScoresParamSchema,
  toolFunctionDefinitionSchema,
  toolsSchema,
  uploadStatusSchema,
  useFunctionSchema,
  userSchema,
  validRuntimesEnum,
  versionSchema,
  viewDataSchema,
  viewDataSearchSchema,
  viewOptionsSchema,
  viewSchema,
  viewTypeEnum
});
