// braintrust/dist/auto-instrumentations/bundler/webpack.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/auto-instrumentations/bundler/webpack.ts
var webpack_exports = {};
__export(webpack_exports, {
  webpackPlugin: () => webpackPlugin
});
module.exports = __toCommonJS(webpack_exports);

// src/auto-instrumentations/bundler/plugin.ts
var import_unplugin = require("unplugin");
var import_code_transformer = require("@apm-js-collab/code-transformer");
var import_path = require("path");
var import_fs = require("fs");
var import_url = require("url");
var import_module_details_from_path = __toESM(require("module-details-from-path"));

// src/auto-instrumentations/configs/openai.ts
var openaiConfigs = [
  // Chat Completions
  {
    channelName: "chat.completions.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "create",
      kind: "Async"
    }
  },
  // Embeddings
  {
    channelName: "embeddings.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/embeddings.mjs"
    },
    functionQuery: {
      className: "Embeddings",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Chat Completions Parse
  {
    channelName: "beta.chat.completions.parse",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/beta/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "parse",
      kind: "Async"
    }
  },
  // Moderations
  {
    channelName: "moderations.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/moderations.mjs"
    },
    functionQuery: {
      className: "Moderations",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Chat Completions Stream
  {
    channelName: "beta.chat.completions.stream",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/beta/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "stream",
      kind: "Sync"
    }
  },
  // Responses API (v4.87.0+)
  {
    channelName: "responses.create",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "create",
      kind: "Async"
    }
  },
  {
    channelName: "responses.stream",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "stream",
      kind: "Sync"
    }
  },
  {
    channelName: "responses.parse",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "parse",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/anthropic.ts
var anthropicConfigs = [
  // Messages API - create (supports streaming via stream=true parameter)
  {
    channelName: "messages.create",
    module: {
      name: "@anthropic-ai/sdk",
      versionRange: ">=0.60.0",
      filePath: "resources/messages.mjs"
    },
    functionQuery: {
      className: "Messages",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Messages API - create (supports streaming via stream=true parameter)
  {
    channelName: "beta.messages.create",
    module: {
      name: "@anthropic-ai/sdk",
      versionRange: ">=0.60.0",
      filePath: "resources/beta/messages/messages.mjs"
    },
    functionQuery: {
      className: "Messages",
      methodName: "create",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/ai-sdk.ts
var aiSDKConfigs = [
  // generateText - async function
  {
    channelName: "generateText",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "generateText",
      kind: "Async"
    }
  },
  // streamText - async function
  {
    channelName: "streamText",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "streamText",
      kind: "Async"
    }
  },
  // generateObject - async function
  {
    channelName: "generateObject",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "generateObject",
      kind: "Async"
    }
  },
  // streamObject - async function
  {
    channelName: "streamObject",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "streamObject",
      kind: "Async"
    }
  },
  // Agent.generate - async method (v3-v5 only, Agent structure changed in v6)
  {
    channelName: "Agent.generate",
    module: {
      name: "ai",
      versionRange: ">=3.0.0 <6.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      className: "Agent",
      methodName: "generate",
      kind: "Async"
    }
  },
  // Agent.stream - async method (v3-v5 only, Agent structure changed in v6)
  {
    channelName: "Agent.stream",
    module: {
      name: "ai",
      versionRange: ">=3.0.0 <6.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      className: "Agent",
      methodName: "stream",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/claude-agent-sdk.ts
var claudeAgentSDKConfigs = [
  // query - Main entry point for agent interactions (top-level exported async generator function)
  {
    channelName: "query",
    module: {
      name: "@anthropic-ai/claude-agent-sdk",
      versionRange: ">=0.1.0",
      filePath: "sdk.mjs"
    },
    functionQuery: {
      functionName: "query",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/google-genai.ts
var googleGenAIConfigs = [
  // Models.generateContentInternal - The actual class method (Node.js entry point)
  // Note: generateContent is an arrow function property that calls this internal method
  {
    channelName: "models.generateContent",
    module: {
      name: "@google/genai",
      versionRange: ">=1.0.0",
      filePath: "dist/node/index.mjs"
    },
    functionQuery: {
      className: "Models",
      methodName: "generateContentInternal",
      kind: "Async"
    }
  },
  // Models.generateContentStreamInternal - The actual class method (Node.js entry point)
  // Note: generateContentStream is an arrow function property that calls this internal method
  {
    channelName: "models.generateContentStream",
    module: {
      name: "@google/genai",
      versionRange: ">=1.0.0",
      filePath: "dist/node/index.mjs"
    },
    functionQuery: {
      className: "Models",
      methodName: "generateContentStreamInternal",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/bundler/plugin.ts
function getModuleVersion(basedir) {
  try {
    const packageJsonPath = (0, import_path.join)(basedir, "package.json");
    const packageJson = JSON.parse((0, import_fs.readFileSync)(packageJsonPath, "utf8"));
    if (packageJson.version) {
      return packageJson.version;
    }
  } catch (error) {
  }
  return void 0;
}
var unplugin = (0, import_unplugin.createUnplugin)((options = {}) => {
  const allInstrumentations = [
    ...openaiConfigs,
    ...anthropicConfigs,
    ...aiSDKConfigs,
    ...claudeAgentSDKConfigs,
    ...googleGenAIConfigs,
    ...options.instrumentations || []
  ];
  const dcModule = options.browser === false ? void 0 : "dc-browser";
  const instrumentationMatcher = (0, import_code_transformer.create)(allInstrumentations, dcModule);
  return {
    name: "code-transformer",
    enforce: "pre",
    transform(code, id) {
      const filePath = id.startsWith("file:") ? (0, import_url.fileURLToPath)(id) : id;
      const ext = (0, import_path.extname)(filePath);
      let isModule = ext === ".mjs" || ext === ".ts" || ext === ".tsx";
      if (ext === ".js") {
        isModule = code.includes("export ") || code.includes("import ");
      }
      const normalizedForPlatform = filePath.split("/").join(import_path.sep);
      const moduleDetails = (0, import_module_details_from_path.default)(normalizedForPlatform);
      if (!moduleDetails) {
        return null;
      }
      const moduleName = moduleDetails.name;
      const moduleVersion = getModuleVersion(moduleDetails.basedir);
      if (!moduleVersion) {
        console.warn(
          `No 'package.json' version found for module ${moduleName} at ${moduleDetails.basedir}. Skipping transformation.`
        );
        return null;
      }
      const normalizedModulePath = moduleDetails.path.replace(/\\/g, "/");
      const transformer = instrumentationMatcher.getTransformer(
        moduleName,
        moduleVersion,
        normalizedModulePath
      );
      if (!transformer) {
        return null;
      }
      try {
        const moduleType = isModule ? "esm" : "cjs";
        const result = transformer.transform(code, moduleType);
        return {
          code: result.code,
          map: result.map
        };
      } catch (error) {
        console.warn(`Code transformation failed for ${id}: ${error}`);
        return null;
      }
    }
  };
});

// src/auto-instrumentations/bundler/webpack.ts
var webpackPlugin = unplugin.webpack;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  webpackPlugin
});
