// dc-browser/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Channel: () => Channel,
  TracingChannel: () => TracingChannel,
  channel: () => channel,
  hasSubscribers: () => hasSubscribers,
  subscribe: () => subscribe,
  tracingChannel: () => tracingChannel,
  unsubscribe: () => unsubscribe
});
module.exports = __toCommonJS(index_exports);

// src/channel.ts
function triggerUncaughtException(err, _) {
  window.dispatchEvent(
    new ErrorEvent("error", {
      error: err,
      message: err?.message
    })
  );
}
function defaultTransform(message) {
  return message;
}
function wrapStoreRun(store, message, next, transform = defaultTransform) {
  return () => {
    let context;
    try {
      context = transform(message);
    } catch (err) {
      queueMicrotask(() => {
        triggerUncaughtException(err, false);
      });
      return next();
    }
    return store.run(context, next);
  };
}
var Channel = class {
  name;
  _subscribers;
  _stores;
  constructor(name) {
    this.name = name;
    this._subscribers = [];
    this._stores = /* @__PURE__ */ new Map();
  }
  get hasSubscribers() {
    return this._subscribers.length > 0;
  }
  subscribe(subscription) {
    if (typeof subscription !== "function") {
      throw new TypeError("subscription must be a function");
    }
    this._subscribers = this._subscribers.slice();
    this._subscribers.push(subscription);
  }
  unsubscribe(subscription) {
    const index = this._subscribers.indexOf(subscription);
    if (index === -1) {
      return false;
    }
    const before = this._subscribers.slice(0, index);
    const after = this._subscribers.slice(index + 1);
    this._subscribers = before;
    this._subscribers.push(...after);
    return true;
  }
  bindStore(store, transform) {
    if (!store || typeof store.run !== "function") {
      throw new TypeError("store must have a run method");
    }
    this._stores.set(store, transform);
  }
  unbindStore(store) {
    if (!this._stores.has(store)) {
      return false;
    }
    this._stores.delete(store);
    return true;
  }
  publish(message) {
    const subscribers = this._subscribers;
    for (let i = 0; i < (subscribers?.length || 0); i++) {
      try {
        const onMessage = subscribers[i];
        onMessage(message, this.name);
      } catch (err) {
        queueMicrotask(() => {
          triggerUncaughtException(err, false);
        });
      }
    }
  }
  runStores(message, fn, thisArg, ...args) {
    let run = () => {
      this.publish(message);
      return Reflect.apply(fn, thisArg, args);
    };
    for (const entry of this._stores.entries()) {
      const store = entry[0];
      const transform = entry[1];
      run = wrapStoreRun(store, message, run, transform);
    }
    return run();
  }
};

// src/channel-registry.ts
var CHANNELS_KEY = /* @__PURE__ */ Symbol.for("dc-browser:channels");
var channels = globalThis[CHANNELS_KEY] || (globalThis[CHANNELS_KEY] = /* @__PURE__ */ new Map());
function getOrCreateChannel(name) {
  let ch = channels.get(name);
  if (!ch) {
    ch = new Channel(name);
    channels.set(name, ch);
  }
  return ch;
}
function getChannelIfExists(name) {
  return channels.get(name);
}

// src/channel-api.ts
function channel(name) {
  if (typeof name !== "string" && typeof name !== "symbol") {
    throw new TypeError("channel name must be a string or symbol");
  }
  return getOrCreateChannel(name);
}
function hasSubscribers(name) {
  const ch = getChannelIfExists(name);
  return ch ? ch.hasSubscribers : false;
}

// src/tracing-channel.ts
var TracingChannel = class {
  start;
  end;
  asyncStart;
  asyncEnd;
  error;
  constructor(nameOrChannels) {
    if (typeof nameOrChannels === "string") {
      this.start = channel(`tracing:${nameOrChannels}:start`);
      this.end = channel(`tracing:${nameOrChannels}:end`);
      this.asyncStart = channel(`tracing:${nameOrChannels}:asyncStart`);
      this.asyncEnd = channel(`tracing:${nameOrChannels}:asyncEnd`);
      this.error = channel(`tracing:${nameOrChannels}:error`);
    } else {
      this.start = nameOrChannels.start;
      this.end = nameOrChannels.end;
      this.asyncStart = nameOrChannels.asyncStart;
      this.asyncEnd = nameOrChannels.asyncEnd;
      this.error = nameOrChannels.error;
    }
  }
  get hasSubscribers() {
    return this.start?.hasSubscribers || this.end?.hasSubscribers || this.asyncStart?.hasSubscribers || this.asyncEnd?.hasSubscribers || this.error?.hasSubscribers || false;
  }
  subscribe(handlers) {
    if (handlers.start) {
      this.start?.subscribe(handlers.start);
    }
    if (handlers.end) {
      this.end?.subscribe(handlers.end);
    }
    if (handlers.asyncStart) {
      this.asyncStart?.subscribe(handlers.asyncStart);
    }
    if (handlers.asyncEnd) {
      this.asyncEnd?.subscribe(handlers.asyncEnd);
    }
    if (handlers.error) {
      this.error?.subscribe(handlers.error);
    }
  }
  unsubscribe(handlers) {
    let done = true;
    if (handlers.start && !this.start?.unsubscribe(handlers.start)) {
      done = false;
    }
    if (handlers.end && !this.end?.unsubscribe(handlers.end)) {
      done = false;
    }
    if (handlers.asyncStart && !this.asyncStart?.unsubscribe(handlers.asyncStart)) {
      done = false;
    }
    if (handlers.asyncEnd && !this.asyncEnd?.unsubscribe(handlers.asyncEnd)) {
      done = false;
    }
    if (handlers.error && !this.error?.unsubscribe(handlers.error)) {
      done = false;
    }
    return done;
  }
  #maybeRunStores(channel2, message, fn, thisArg, ...args) {
    if (!channel2) {
      return Reflect.apply(fn, thisArg, args);
    }
    return channel2.runStores(message, fn, thisArg, ...args);
  }
  traceSync(fn, message, thisArg, ...args) {
    if (!this.hasSubscribers) {
      return Reflect.apply(fn, thisArg, args);
    }
    return this.#maybeRunStores(
      this.start,
      message,
      () => {
        try {
          const result = Reflect.apply(fn, thisArg, args);
          message.result = result;
          return result;
        } catch (err) {
          message.error = err;
          this.error?.publish(message);
          throw err;
        } finally {
          this.end?.publish(message);
        }
      }
    );
  }
  tracePromise(fn, message, thisArg, ...args) {
    if (!this.hasSubscribers) {
      return Reflect.apply(fn, thisArg, args);
    }
    const { start, end, asyncStart, asyncEnd, error } = this;
    function reject(err) {
      message.error = err;
      error?.publish(message);
      asyncStart?.publish(message);
      asyncEnd?.publish(message);
      return Promise.reject(err);
    }
    function resolve(result) {
      message.result = result;
      asyncStart?.publish(message);
      asyncEnd?.publish(message);
      return result;
    }
    return this.#maybeRunStores(
      start,
      message,
      () => {
        try {
          let promise = Reflect.apply(fn, thisArg, args);
          if (!(promise instanceof Promise)) {
            promise = Promise.resolve(promise);
          }
          return promise.then(resolve, reject);
        } catch (err) {
          message.error = err;
          error?.publish(message);
          throw err;
        } finally {
          end?.publish(message);
        }
      }
    );
  }
  traceCallback(fn, position = -1, message, thisArg, ...args) {
    if (!this.hasSubscribers) {
      return Reflect.apply(fn, thisArg, args);
    }
    const { start, end, asyncStart, asyncEnd, error } = this;
    const self = this;
    function wrappedCallback(err, res) {
      if (err) {
        message.error = err;
        error?.publish(message);
      } else {
        message.result = res;
      }
      self.#maybeRunStores(
        asyncStart,
        message,
        () => {
          try {
            return Reflect.apply(callback, thisArg, arguments);
          } finally {
            asyncEnd?.publish(message);
          }
        }
      );
    }
    const callback = args.at(position);
    if (typeof callback !== "function") {
      throw new TypeError("callback must be a function");
    }
    args.splice(position, 1, wrappedCallback);
    return this.#maybeRunStores(
      start,
      message,
      () => {
        try {
          return Reflect.apply(fn, thisArg, args);
        } catch (err) {
          message.error = err;
          error?.publish(message);
          throw err;
        } finally {
          end?.publish(message);
        }
      }
    );
  }
};

// src/index.ts
function subscribe(name, subscription) {
  const ch = channel(name);
  ch.subscribe(subscription);
}
function unsubscribe(name, subscription) {
  const ch = channel(name);
  return ch.unsubscribe(subscription);
}
function tracingChannel(nameOrChannels) {
  return new TracingChannel(nameOrChannels);
}
