// braintrust/dist/auto-instrumentations/chunk-KVX7OFPD.mjs
// src/auto-instrumentations/configs/openai.ts
var openaiConfigs = [
  // Chat Completions
  {
    channelName: "chat.completions.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "create",
      kind: "Async"
    }
  },
  // Embeddings
  {
    channelName: "embeddings.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/embeddings.mjs"
    },
    functionQuery: {
      className: "Embeddings",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Chat Completions Parse
  {
    channelName: "beta.chat.completions.parse",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/beta/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "parse",
      kind: "Async"
    }
  },
  // Moderations
  {
    channelName: "moderations.create",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/moderations.mjs"
    },
    functionQuery: {
      className: "Moderations",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Chat Completions Stream
  {
    channelName: "beta.chat.completions.stream",
    module: {
      name: "openai",
      versionRange: ">=4.0.0",
      filePath: "resources/beta/chat/completions.mjs"
    },
    functionQuery: {
      className: "Completions",
      methodName: "stream",
      kind: "Sync"
    }
  },
  // Responses API (v4.87.0+)
  {
    channelName: "responses.create",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "create",
      kind: "Async"
    }
  },
  {
    channelName: "responses.stream",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "stream",
      kind: "Sync"
    }
  },
  {
    channelName: "responses.parse",
    module: {
      name: "openai",
      versionRange: ">=4.87.0",
      filePath: "resources/responses/responses.mjs"
    },
    functionQuery: {
      className: "Responses",
      methodName: "parse",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/anthropic.ts
var anthropicConfigs = [
  // Messages API - create (supports streaming via stream=true parameter)
  {
    channelName: "messages.create",
    module: {
      name: "@anthropic-ai/sdk",
      versionRange: ">=0.60.0",
      filePath: "resources/messages.mjs"
    },
    functionQuery: {
      className: "Messages",
      methodName: "create",
      kind: "Async"
    }
  },
  // Beta Messages API - create (supports streaming via stream=true parameter)
  {
    channelName: "beta.messages.create",
    module: {
      name: "@anthropic-ai/sdk",
      versionRange: ">=0.60.0",
      filePath: "resources/beta/messages/messages.mjs"
    },
    functionQuery: {
      className: "Messages",
      methodName: "create",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/ai-sdk.ts
var aiSDKConfigs = [
  // generateText - async function
  {
    channelName: "generateText",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "generateText",
      kind: "Async"
    }
  },
  // streamText - async function
  {
    channelName: "streamText",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "streamText",
      kind: "Async"
    }
  },
  // generateObject - async function
  {
    channelName: "generateObject",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "generateObject",
      kind: "Async"
    }
  },
  // streamObject - async function
  {
    channelName: "streamObject",
    module: {
      name: "ai",
      versionRange: ">=3.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      functionName: "streamObject",
      kind: "Async"
    }
  },
  // Agent.generate - async method (v3-v5 only, Agent structure changed in v6)
  {
    channelName: "Agent.generate",
    module: {
      name: "ai",
      versionRange: ">=3.0.0 <6.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      className: "Agent",
      methodName: "generate",
      kind: "Async"
    }
  },
  // Agent.stream - async method (v3-v5 only, Agent structure changed in v6)
  {
    channelName: "Agent.stream",
    module: {
      name: "ai",
      versionRange: ">=3.0.0 <6.0.0",
      filePath: "dist/index.mjs"
    },
    functionQuery: {
      className: "Agent",
      methodName: "stream",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/claude-agent-sdk.ts
var claudeAgentSDKConfigs = [
  // query - Main entry point for agent interactions (top-level exported async generator function)
  {
    channelName: "query",
    module: {
      name: "@anthropic-ai/claude-agent-sdk",
      versionRange: ">=0.1.0",
      filePath: "sdk.mjs"
    },
    functionQuery: {
      functionName: "query",
      kind: "Async"
    }
  }
];

// src/auto-instrumentations/configs/google-genai.ts
var googleGenAIConfigs = [
  // Models.generateContentInternal - The actual class method (Node.js entry point)
  // Note: generateContent is an arrow function property that calls this internal method
  {
    channelName: "models.generateContent",
    module: {
      name: "@google/genai",
      versionRange: ">=1.0.0",
      filePath: "dist/node/index.mjs"
    },
    functionQuery: {
      className: "Models",
      methodName: "generateContentInternal",
      kind: "Async"
    }
  },
  // Models.generateContentStreamInternal - The actual class method (Node.js entry point)
  // Note: generateContentStream is an arrow function property that calls this internal method
  {
    channelName: "models.generateContentStream",
    module: {
      name: "@google/genai",
      versionRange: ">=1.0.0",
      filePath: "dist/node/index.mjs"
    },
    functionQuery: {
      className: "Models",
      methodName: "generateContentStreamInternal",
      kind: "Async"
    }
  }
];

export {
  openaiConfigs,
  anthropicConfigs,
  aiSDKConfigs,
  claudeAgentSDKConfigs,
  googleGenAIConfigs
};
